{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2beeee7d-19d6-430f-8f0f-4f9e504c9994",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if (!requireNamespace(\"BiocManager\", quietly=TRUE)) \n",
    "#     install.packages(\"BiocManager\") \n",
    "\n",
    "# BiocManager::install(\"ANCOMBC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "712c0b3d-627a-49cf-9714-e6dfd145e0e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.5.2     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.3.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.4     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.1.0     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "Welcome! Want to learn more? See two factoextra-related books at https://goo.gl/ve3WBa\n",
      "\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mThere were 2 warnings in `mutate()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `Age = as.numeric(Age)`.\n",
      "Caused by warning:\n",
      "\u001b[33m!\u001b[39m NAs introduced by coercion\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 1 remaining warning.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mThere were 2 warnings in `mutate()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `Age = as.numeric(Age)`.\n",
      "Caused by warning:\n",
      "\u001b[33m!\u001b[39m NAs introduced by coercion\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 1 remaining warning.”\n"
     ]
    }
   ],
   "source": [
    ".libPaths(R.home('/opt/homebrew/Caskroom/miniforge/base/lib/R/library'))\n",
    "Output = '/Users/alexis/Library/CloudStorage/OneDrive-UniversityofNorthCarolinaatChapelHill/CEMALB_DataAnalysisPM/Projects/P1013. Microbiome Peanut Allergy/P1013.3. Analyses/P1013.3.1. Data Processing/Output'\n",
    "cur_date = '083025'\n",
    "\n",
    "library(readxl)\n",
    "library(openxlsx)\n",
    "library(tidyverse)\n",
    "library(factoextra)\n",
    "library(ANCOMBC) \n",
    "\n",
    "# reading in files\n",
    "saliva_df = data.frame(read_excel(\"Input/Saliva_Diversity_082925.xlsx\", sheet = 2)) %>%\n",
    "    mutate(Age = as.numeric(Age), IgE = as.numeric(IgE))\n",
    "stool_df = data.frame(read_excel(\"Input/Stool_Diversity_082925.xlsx\", sheet = 2)) %>%\n",
    "    mutate(Age = as.numeric(Age), IgE = as.numeric(IgE))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "615998a2-0215-4330-b5cb-b7444aac6ef1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 284</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Original_Sample_ID</th><th scope=col>Subject_No</th><th scope=col>Sample_ID</th><th scope=col>Time_Point</th><th scope=col>Treatment</th><th scope=col>Outcome</th><th scope=col>Age</th><th scope=col>IgE</th><th scope=col>IgE_Binarized</th><th scope=col>k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Peptococcaceae.g__Peptococcus.s__sp._HMT_167</th><th scope=col>⋯</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.s__sp._HMT_219</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.__</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__wadei</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__shahii</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_498</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_215</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_392</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_212</th><th scope=col>k__Bacteria.p__Gracilibacteria_.GN02..c__Gracilibacteria_.GN02._.C.1..o__Gracilibacteria_.GN02._.O.1..f__Gracilibacteria_.GN02._.F.1..g__Gracilibacteria_.GN02._.G.1..s__bacterium_HMT_872</th><th scope=col>k__Bacteria.p__Firmicutes.c__Bacilli.o__Lactobacillales.f__Lactobacillaceae.g__Lacticaseibacillus.__</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>FS01baseline</td><td>1</td><td>Pre_1 </td><td>Pre </td><td>P</td><td>F </td><td>3.838356</td><td>41.00</td><td>High</td><td> 0</td><td>⋯</td><td>0</td><td>  0</td><td>0</td><td>0</td><td>0</td><td> 18</td><td>15</td><td> 42</td><td> 0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>FS0236mo    </td><td>2</td><td>Post_2</td><td>Post</td><td>A</td><td>DS</td><td>1.873973</td><td>14.20</td><td>Low </td><td>59</td><td>⋯</td><td>0</td><td>  0</td><td>0</td><td>0</td><td>0</td><td>532</td><td>30</td><td> 27</td><td> 0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>FS02baseline</td><td>2</td><td>Pre_2 </td><td>Pre </td><td>A</td><td>DS</td><td>1.873973</td><td>14.20</td><td>Low </td><td> 0</td><td>⋯</td><td>0</td><td>198</td><td>0</td><td>0</td><td>0</td><td>172</td><td> 0</td><td> 64</td><td> 0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>FS0436mo    </td><td>3</td><td>Post_3</td><td>Post</td><td>P</td><td>F </td><td>3.715068</td><td>67.40</td><td>High</td><td> 0</td><td>⋯</td><td>0</td><td>  0</td><td>0</td><td>0</td><td>0</td><td>  0</td><td> 0</td><td>  0</td><td>15</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>FS04baseline</td><td>3</td><td>Pre_3 </td><td>Pre </td><td>P</td><td>F </td><td>3.715068</td><td>67.40</td><td>High</td><td> 0</td><td>⋯</td><td>0</td><td>  0</td><td>0</td><td>0</td><td>0</td><td>  0</td><td> 0</td><td>  8</td><td> 0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>FS0536mo    </td><td>4</td><td>Post_4</td><td>Post</td><td>A</td><td>DS</td><td>3.991781</td><td> 9.59</td><td>Low </td><td>24</td><td>⋯</td><td>0</td><td>  0</td><td>0</td><td>0</td><td>0</td><td>676</td><td> 0</td><td>207</td><td> 0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 284\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & Original\\_Sample\\_ID & Subject\\_No & Sample\\_ID & Time\\_Point & Treatment & Outcome & Age & IgE & IgE\\_Binarized & k\\_\\_Bacteria.p\\_\\_Firmicutes.c\\_\\_Clostridia.o\\_\\_Eubacteriales.f\\_\\_Peptococcaceae.g\\_\\_Peptococcus.s\\_\\_sp.\\_HMT\\_167 & ⋯ & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Pseudoleptotrichia.s\\_\\_sp.\\_HMT\\_219 & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Pseudoleptotrichia.\\_\\_ & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Leptotrichia.s\\_\\_wadei & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Leptotrichia.s\\_\\_shahii & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Leptotrichia.s\\_\\_sp.\\_HMT\\_498 & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Leptotrichia.s\\_\\_sp.\\_HMT\\_215 & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Leptotrichia.s\\_\\_sp.\\_HMT\\_392 & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Leptotrichia.s\\_\\_sp.\\_HMT\\_212 & k\\_\\_Bacteria.p\\_\\_Gracilibacteria\\_.GN02..c\\_\\_Gracilibacteria\\_.GN02.\\_.C.1..o\\_\\_Gracilibacteria\\_.GN02.\\_.O.1..f\\_\\_Gracilibacteria\\_.GN02.\\_.F.1..g\\_\\_Gracilibacteria\\_.GN02.\\_.G.1..s\\_\\_bacterium\\_HMT\\_872 & k\\_\\_Bacteria.p\\_\\_Firmicutes.c\\_\\_Bacilli.o\\_\\_Lactobacillales.f\\_\\_Lactobacillaceae.g\\_\\_Lacticaseibacillus.\\_\\_\\\\\n",
       "  & <chr> & <dbl> & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <chr> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & FS01baseline & 1 & Pre\\_1  & Pre  & P & F  & 3.838356 & 41.00 & High &  0 & ⋯ & 0 &   0 & 0 & 0 & 0 &  18 & 15 &  42 &  0 & 0\\\\\n",
       "\t2 & FS0236mo     & 2 & Post\\_2 & Post & A & DS & 1.873973 & 14.20 & Low  & 59 & ⋯ & 0 &   0 & 0 & 0 & 0 & 532 & 30 &  27 &  0 & 0\\\\\n",
       "\t3 & FS02baseline & 2 & Pre\\_2  & Pre  & A & DS & 1.873973 & 14.20 & Low  &  0 & ⋯ & 0 & 198 & 0 & 0 & 0 & 172 &  0 &  64 &  0 & 0\\\\\n",
       "\t4 & FS0436mo     & 3 & Post\\_3 & Post & P & F  & 3.715068 & 67.40 & High &  0 & ⋯ & 0 &   0 & 0 & 0 & 0 &   0 &  0 &   0 & 15 & 0\\\\\n",
       "\t5 & FS04baseline & 3 & Pre\\_3  & Pre  & P & F  & 3.715068 & 67.40 & High &  0 & ⋯ & 0 &   0 & 0 & 0 & 0 &   0 &  0 &   8 &  0 & 0\\\\\n",
       "\t6 & FS0536mo     & 4 & Post\\_4 & Post & A & DS & 3.991781 &  9.59 & Low  & 24 & ⋯ & 0 &   0 & 0 & 0 & 0 & 676 &  0 & 207 &  0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 284\n",
       "\n",
       "| <!--/--> | Original_Sample_ID &lt;chr&gt; | Subject_No &lt;dbl&gt; | Sample_ID &lt;chr&gt; | Time_Point &lt;chr&gt; | Treatment &lt;chr&gt; | Outcome &lt;chr&gt; | Age &lt;dbl&gt; | IgE &lt;dbl&gt; | IgE_Binarized &lt;chr&gt; | k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Peptococcaceae.g__Peptococcus.s__sp._HMT_167 &lt;dbl&gt; | ⋯ ⋯ | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.s__sp._HMT_219 &lt;dbl&gt; | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.__ &lt;dbl&gt; | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__wadei &lt;dbl&gt; | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__shahii &lt;dbl&gt; | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_498 &lt;dbl&gt; | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_215 &lt;dbl&gt; | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_392 &lt;dbl&gt; | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_212 &lt;dbl&gt; | k__Bacteria.p__Gracilibacteria_.GN02..c__Gracilibacteria_.GN02._.C.1..o__Gracilibacteria_.GN02._.O.1..f__Gracilibacteria_.GN02._.F.1..g__Gracilibacteria_.GN02._.G.1..s__bacterium_HMT_872 &lt;dbl&gt; | k__Bacteria.p__Firmicutes.c__Bacilli.o__Lactobacillales.f__Lactobacillaceae.g__Lacticaseibacillus.__ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | FS01baseline | 1 | Pre_1  | Pre  | P | F  | 3.838356 | 41.00 | High |  0 | ⋯ | 0 |   0 | 0 | 0 | 0 |  18 | 15 |  42 |  0 | 0 |\n",
       "| 2 | FS0236mo     | 2 | Post_2 | Post | A | DS | 1.873973 | 14.20 | Low  | 59 | ⋯ | 0 |   0 | 0 | 0 | 0 | 532 | 30 |  27 |  0 | 0 |\n",
       "| 3 | FS02baseline | 2 | Pre_2  | Pre  | A | DS | 1.873973 | 14.20 | Low  |  0 | ⋯ | 0 | 198 | 0 | 0 | 0 | 172 |  0 |  64 |  0 | 0 |\n",
       "| 4 | FS0436mo     | 3 | Post_3 | Post | P | F  | 3.715068 | 67.40 | High |  0 | ⋯ | 0 |   0 | 0 | 0 | 0 |   0 |  0 |   0 | 15 | 0 |\n",
       "| 5 | FS04baseline | 3 | Pre_3  | Pre  | P | F  | 3.715068 | 67.40 | High |  0 | ⋯ | 0 |   0 | 0 | 0 | 0 |   0 |  0 |   8 |  0 | 0 |\n",
       "| 6 | FS0536mo     | 4 | Post_4 | Post | A | DS | 3.991781 |  9.59 | Low  | 24 | ⋯ | 0 |   0 | 0 | 0 | 0 | 676 |  0 | 207 |  0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  Original_Sample_ID Subject_No Sample_ID Time_Point Treatment Outcome Age     \n",
       "1 FS01baseline       1          Pre_1     Pre        P         F       3.838356\n",
       "2 FS0236mo           2          Post_2    Post       A         DS      1.873973\n",
       "3 FS02baseline       2          Pre_2     Pre        A         DS      1.873973\n",
       "4 FS0436mo           3          Post_3    Post       P         F       3.715068\n",
       "5 FS04baseline       3          Pre_3     Pre        P         F       3.715068\n",
       "6 FS0536mo           4          Post_4    Post       A         DS      3.991781\n",
       "  IgE   IgE_Binarized\n",
       "1 41.00 High         \n",
       "2 14.20 Low          \n",
       "3 14.20 Low          \n",
       "4 67.40 High         \n",
       "5 67.40 High         \n",
       "6  9.59 Low          \n",
       "  k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Peptococcaceae.g__Peptococcus.s__sp._HMT_167\n",
       "1  0                                                                                                      \n",
       "2 59                                                                                                      \n",
       "3  0                                                                                                      \n",
       "4  0                                                                                                      \n",
       "5  0                                                                                                      \n",
       "6 24                                                                                                      \n",
       "  ⋯\n",
       "1 ⋯\n",
       "2 ⋯\n",
       "3 ⋯\n",
       "4 ⋯\n",
       "5 ⋯\n",
       "6 ⋯\n",
       "  k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.s__sp._HMT_219\n",
       "1 0                                                                                                                       \n",
       "2 0                                                                                                                       \n",
       "3 0                                                                                                                       \n",
       "4 0                                                                                                                       \n",
       "5 0                                                                                                                       \n",
       "6 0                                                                                                                       \n",
       "  k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.__\n",
       "1   0                                                                                                         \n",
       "2   0                                                                                                         \n",
       "3 198                                                                                                         \n",
       "4   0                                                                                                         \n",
       "5   0                                                                                                         \n",
       "6   0                                                                                                         \n",
       "  k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__wadei\n",
       "1 0                                                                                                           \n",
       "2 0                                                                                                           \n",
       "3 0                                                                                                           \n",
       "4 0                                                                                                           \n",
       "5 0                                                                                                           \n",
       "6 0                                                                                                           \n",
       "  k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__shahii\n",
       "1 0                                                                                                            \n",
       "2 0                                                                                                            \n",
       "3 0                                                                                                            \n",
       "4 0                                                                                                            \n",
       "5 0                                                                                                            \n",
       "6 0                                                                                                            \n",
       "  k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_498\n",
       "1 0                                                                                                                 \n",
       "2 0                                                                                                                 \n",
       "3 0                                                                                                                 \n",
       "4 0                                                                                                                 \n",
       "5 0                                                                                                                 \n",
       "6 0                                                                                                                 \n",
       "  k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_215\n",
       "1  18                                                                                                               \n",
       "2 532                                                                                                               \n",
       "3 172                                                                                                               \n",
       "4   0                                                                                                               \n",
       "5   0                                                                                                               \n",
       "6 676                                                                                                               \n",
       "  k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_392\n",
       "1 15                                                                                                                \n",
       "2 30                                                                                                                \n",
       "3  0                                                                                                                \n",
       "4  0                                                                                                                \n",
       "5  0                                                                                                                \n",
       "6  0                                                                                                                \n",
       "  k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_212\n",
       "1  42                                                                                                               \n",
       "2  27                                                                                                               \n",
       "3  64                                                                                                               \n",
       "4   0                                                                                                               \n",
       "5   8                                                                                                               \n",
       "6 207                                                                                                               \n",
       "  k__Bacteria.p__Gracilibacteria_.GN02..c__Gracilibacteria_.GN02._.C.1..o__Gracilibacteria_.GN02._.O.1..f__Gracilibacteria_.GN02._.F.1..g__Gracilibacteria_.GN02._.G.1..s__bacterium_HMT_872\n",
       "1  0                                                                                                                                                                                        \n",
       "2  0                                                                                                                                                                                        \n",
       "3  0                                                                                                                                                                                        \n",
       "4 15                                                                                                                                                                                        \n",
       "5  0                                                                                                                                                                                        \n",
       "6  0                                                                                                                                                                                        \n",
       "  k__Bacteria.p__Firmicutes.c__Bacilli.o__Lactobacillales.f__Lactobacillaceae.g__Lacticaseibacillus.__\n",
       "1 0                                                                                                   \n",
       "2 0                                                                                                   \n",
       "3 0                                                                                                   \n",
       "4 0                                                                                                   \n",
       "5 0                                                                                                   \n",
       "6 0                                                                                                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(saliva_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66865086-cf70-4a70-8895-ef7bf834a903",
   "metadata": {},
   "source": [
    "Running individual statistics on saliva and stool microbiome samples at baseline to see what species differ across outcomes and if age plays a role. I used the [ANCOM-BC2](https://www.bioconductor.org/packages/release/bioc/vignettes/ANCOMBC/inst/doc/ANCOMBC2.html) package to perform a differential abundance analysis ([ref](https://www.nature.com/articles/s41592-023-02092-7#Sec1)). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "105171ad-0862-4eaa-8a04-e2e88af3e5cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 283</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Original_Sample_ID</th><th scope=col>Subject_No</th><th scope=col>Time_Point</th><th scope=col>Treatment</th><th scope=col>Outcome</th><th scope=col>Age</th><th scope=col>IgE</th><th scope=col>IgE_Binarized</th><th scope=col>k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Peptococcaceae.g__Peptococcus.s__sp._HMT_167</th><th scope=col>k__Bacteria.p__Actinobacteria.c__Actinomycetia.o__Propionibacteriales.f__Propionibacteriaceae.g__Arachnia.__</th><th scope=col>⋯</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.s__sp._HMT_219</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.__</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__wadei</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__shahii</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_498</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_215</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_392</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_212</th><th scope=col>k__Bacteria.p__Gracilibacteria_.GN02..c__Gracilibacteria_.GN02._.C.1..o__Gracilibacteria_.GN02._.O.1..f__Gracilibacteria_.GN02._.F.1..g__Gracilibacteria_.GN02._.G.1..s__bacterium_HMT_872</th><th scope=col>k__Bacteria.p__Firmicutes.c__Bacilli.o__Lactobacillales.f__Lactobacillaceae.g__Lacticaseibacillus.__</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>Pre_1</th><td>FS01baseline</td><td>1</td><td>Pre</td><td>P</td><td>F </td><td>3.838356</td><td> 41.00</td><td>High</td><td> 0</td><td> 10</td><td>⋯</td><td>0</td><td>   0</td><td>0</td><td>0</td><td>0</td><td> 18</td><td>15</td><td> 42</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>Pre_2</th><td>FS02baseline</td><td>2</td><td>Pre</td><td>A</td><td>F </td><td>1.873973</td><td> 14.20</td><td>Low </td><td> 0</td><td> 19</td><td>⋯</td><td>0</td><td> 198</td><td>0</td><td>0</td><td>0</td><td>172</td><td> 0</td><td> 64</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>Pre_3</th><td>FS04baseline</td><td>3</td><td>Pre</td><td>P</td><td>F </td><td>3.715068</td><td> 67.40</td><td>High</td><td> 0</td><td>  0</td><td>⋯</td><td>0</td><td>   0</td><td>0</td><td>0</td><td>0</td><td>  0</td><td> 0</td><td>  8</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>Pre_4</th><td>FS05baseline</td><td>4</td><td>Pre</td><td>A</td><td>F </td><td>3.991781</td><td>  9.59</td><td>Low </td><td>51</td><td> 23</td><td>⋯</td><td>0</td><td>   0</td><td>0</td><td>0</td><td>0</td><td> 78</td><td> 0</td><td> 83</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>Pre_5</th><td>FS06baseline</td><td>5</td><td>Pre</td><td>P</td><td>F </td><td>2.356164</td><td>106.00</td><td>High</td><td> 0</td><td>204</td><td>⋯</td><td>0</td><td>   0</td><td>0</td><td>0</td><td>0</td><td> 23</td><td>33</td><td>286</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>Pre_6</th><td>FS07baseline</td><td>6</td><td>Pre</td><td>A</td><td>SU</td><td>1.419178</td><td>  1.32</td><td>Low </td><td> 0</td><td>  0</td><td>⋯</td><td>0</td><td>1798</td><td>0</td><td>0</td><td>0</td><td>192</td><td> 0</td><td>  0</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 283\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & Original\\_Sample\\_ID & Subject\\_No & Time\\_Point & Treatment & Outcome & Age & IgE & IgE\\_Binarized & k\\_\\_Bacteria.p\\_\\_Firmicutes.c\\_\\_Clostridia.o\\_\\_Eubacteriales.f\\_\\_Peptococcaceae.g\\_\\_Peptococcus.s\\_\\_sp.\\_HMT\\_167 & k\\_\\_Bacteria.p\\_\\_Actinobacteria.c\\_\\_Actinomycetia.o\\_\\_Propionibacteriales.f\\_\\_Propionibacteriaceae.g\\_\\_Arachnia.\\_\\_ & ⋯ & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Pseudoleptotrichia.s\\_\\_sp.\\_HMT\\_219 & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Pseudoleptotrichia.\\_\\_ & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Leptotrichia.s\\_\\_wadei & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Leptotrichia.s\\_\\_shahii & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Leptotrichia.s\\_\\_sp.\\_HMT\\_498 & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Leptotrichia.s\\_\\_sp.\\_HMT\\_215 & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Leptotrichia.s\\_\\_sp.\\_HMT\\_392 & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Leptotrichia.s\\_\\_sp.\\_HMT\\_212 & k\\_\\_Bacteria.p\\_\\_Gracilibacteria\\_.GN02..c\\_\\_Gracilibacteria\\_.GN02.\\_.C.1..o\\_\\_Gracilibacteria\\_.GN02.\\_.O.1..f\\_\\_Gracilibacteria\\_.GN02.\\_.F.1..g\\_\\_Gracilibacteria\\_.GN02.\\_.G.1..s\\_\\_bacterium\\_HMT\\_872 & k\\_\\_Bacteria.p\\_\\_Firmicutes.c\\_\\_Bacilli.o\\_\\_Lactobacillales.f\\_\\_Lactobacillaceae.g\\_\\_Lacticaseibacillus.\\_\\_\\\\\n",
       "  & <chr> & <dbl> & <chr> & <chr> & <fct> & <dbl> & <dbl> & <chr> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tPre\\_1 & FS01baseline & 1 & Pre & P & F  & 3.838356 &  41.00 & High &  0 &  10 & ⋯ & 0 &    0 & 0 & 0 & 0 &  18 & 15 &  42 & 0 & 0\\\\\n",
       "\tPre\\_2 & FS02baseline & 2 & Pre & A & F  & 1.873973 &  14.20 & Low  &  0 &  19 & ⋯ & 0 &  198 & 0 & 0 & 0 & 172 &  0 &  64 & 0 & 0\\\\\n",
       "\tPre\\_3 & FS04baseline & 3 & Pre & P & F  & 3.715068 &  67.40 & High &  0 &   0 & ⋯ & 0 &    0 & 0 & 0 & 0 &   0 &  0 &   8 & 0 & 0\\\\\n",
       "\tPre\\_4 & FS05baseline & 4 & Pre & A & F  & 3.991781 &   9.59 & Low  & 51 &  23 & ⋯ & 0 &    0 & 0 & 0 & 0 &  78 &  0 &  83 & 0 & 0\\\\\n",
       "\tPre\\_5 & FS06baseline & 5 & Pre & P & F  & 2.356164 & 106.00 & High &  0 & 204 & ⋯ & 0 &    0 & 0 & 0 & 0 &  23 & 33 & 286 & 0 & 0\\\\\n",
       "\tPre\\_6 & FS07baseline & 6 & Pre & A & SU & 1.419178 &   1.32 & Low  &  0 &   0 & ⋯ & 0 & 1798 & 0 & 0 & 0 & 192 &  0 &   0 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 283\n",
       "\n",
       "| <!--/--> | Original_Sample_ID &lt;chr&gt; | Subject_No &lt;dbl&gt; | Time_Point &lt;chr&gt; | Treatment &lt;chr&gt; | Outcome &lt;fct&gt; | Age &lt;dbl&gt; | IgE &lt;dbl&gt; | IgE_Binarized &lt;chr&gt; | k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Peptococcaceae.g__Peptococcus.s__sp._HMT_167 &lt;dbl&gt; | k__Bacteria.p__Actinobacteria.c__Actinomycetia.o__Propionibacteriales.f__Propionibacteriaceae.g__Arachnia.__ &lt;dbl&gt; | ⋯ ⋯ | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.s__sp._HMT_219 &lt;dbl&gt; | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.__ &lt;dbl&gt; | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__wadei &lt;dbl&gt; | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__shahii &lt;dbl&gt; | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_498 &lt;dbl&gt; | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_215 &lt;dbl&gt; | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_392 &lt;dbl&gt; | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_212 &lt;dbl&gt; | k__Bacteria.p__Gracilibacteria_.GN02..c__Gracilibacteria_.GN02._.C.1..o__Gracilibacteria_.GN02._.O.1..f__Gracilibacteria_.GN02._.F.1..g__Gracilibacteria_.GN02._.G.1..s__bacterium_HMT_872 &lt;dbl&gt; | k__Bacteria.p__Firmicutes.c__Bacilli.o__Lactobacillales.f__Lactobacillaceae.g__Lacticaseibacillus.__ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Pre_1 | FS01baseline | 1 | Pre | P | F  | 3.838356 |  41.00 | High |  0 |  10 | ⋯ | 0 |    0 | 0 | 0 | 0 |  18 | 15 |  42 | 0 | 0 |\n",
       "| Pre_2 | FS02baseline | 2 | Pre | A | F  | 1.873973 |  14.20 | Low  |  0 |  19 | ⋯ | 0 |  198 | 0 | 0 | 0 | 172 |  0 |  64 | 0 | 0 |\n",
       "| Pre_3 | FS04baseline | 3 | Pre | P | F  | 3.715068 |  67.40 | High |  0 |   0 | ⋯ | 0 |    0 | 0 | 0 | 0 |   0 |  0 |   8 | 0 | 0 |\n",
       "| Pre_4 | FS05baseline | 4 | Pre | A | F  | 3.991781 |   9.59 | Low  | 51 |  23 | ⋯ | 0 |    0 | 0 | 0 | 0 |  78 |  0 |  83 | 0 | 0 |\n",
       "| Pre_5 | FS06baseline | 5 | Pre | P | F  | 2.356164 | 106.00 | High |  0 | 204 | ⋯ | 0 |    0 | 0 | 0 | 0 |  23 | 33 | 286 | 0 | 0 |\n",
       "| Pre_6 | FS07baseline | 6 | Pre | A | SU | 1.419178 |   1.32 | Low  |  0 |   0 | ⋯ | 0 | 1798 | 0 | 0 | 0 | 192 |  0 |   0 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "      Original_Sample_ID Subject_No Time_Point Treatment Outcome Age     \n",
       "Pre_1 FS01baseline       1          Pre        P         F       3.838356\n",
       "Pre_2 FS02baseline       2          Pre        A         F       1.873973\n",
       "Pre_3 FS04baseline       3          Pre        P         F       3.715068\n",
       "Pre_4 FS05baseline       4          Pre        A         F       3.991781\n",
       "Pre_5 FS06baseline       5          Pre        P         F       2.356164\n",
       "Pre_6 FS07baseline       6          Pre        A         SU      1.419178\n",
       "      IgE    IgE_Binarized\n",
       "Pre_1  41.00 High         \n",
       "Pre_2  14.20 Low          \n",
       "Pre_3  67.40 High         \n",
       "Pre_4   9.59 Low          \n",
       "Pre_5 106.00 High         \n",
       "Pre_6   1.32 Low          \n",
       "      k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Peptococcaceae.g__Peptococcus.s__sp._HMT_167\n",
       "Pre_1  0                                                                                                      \n",
       "Pre_2  0                                                                                                      \n",
       "Pre_3  0                                                                                                      \n",
       "Pre_4 51                                                                                                      \n",
       "Pre_5  0                                                                                                      \n",
       "Pre_6  0                                                                                                      \n",
       "      k__Bacteria.p__Actinobacteria.c__Actinomycetia.o__Propionibacteriales.f__Propionibacteriaceae.g__Arachnia.__\n",
       "Pre_1  10                                                                                                         \n",
       "Pre_2  19                                                                                                         \n",
       "Pre_3   0                                                                                                         \n",
       "Pre_4  23                                                                                                         \n",
       "Pre_5 204                                                                                                         \n",
       "Pre_6   0                                                                                                         \n",
       "      ⋯\n",
       "Pre_1 ⋯\n",
       "Pre_2 ⋯\n",
       "Pre_3 ⋯\n",
       "Pre_4 ⋯\n",
       "Pre_5 ⋯\n",
       "Pre_6 ⋯\n",
       "      k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.s__sp._HMT_219\n",
       "Pre_1 0                                                                                                                       \n",
       "Pre_2 0                                                                                                                       \n",
       "Pre_3 0                                                                                                                       \n",
       "Pre_4 0                                                                                                                       \n",
       "Pre_5 0                                                                                                                       \n",
       "Pre_6 0                                                                                                                       \n",
       "      k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.__\n",
       "Pre_1    0                                                                                                        \n",
       "Pre_2  198                                                                                                        \n",
       "Pre_3    0                                                                                                        \n",
       "Pre_4    0                                                                                                        \n",
       "Pre_5    0                                                                                                        \n",
       "Pre_6 1798                                                                                                        \n",
       "      k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__wadei\n",
       "Pre_1 0                                                                                                           \n",
       "Pre_2 0                                                                                                           \n",
       "Pre_3 0                                                                                                           \n",
       "Pre_4 0                                                                                                           \n",
       "Pre_5 0                                                                                                           \n",
       "Pre_6 0                                                                                                           \n",
       "      k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__shahii\n",
       "Pre_1 0                                                                                                            \n",
       "Pre_2 0                                                                                                            \n",
       "Pre_3 0                                                                                                            \n",
       "Pre_4 0                                                                                                            \n",
       "Pre_5 0                                                                                                            \n",
       "Pre_6 0                                                                                                            \n",
       "      k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_498\n",
       "Pre_1 0                                                                                                                 \n",
       "Pre_2 0                                                                                                                 \n",
       "Pre_3 0                                                                                                                 \n",
       "Pre_4 0                                                                                                                 \n",
       "Pre_5 0                                                                                                                 \n",
       "Pre_6 0                                                                                                                 \n",
       "      k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_215\n",
       "Pre_1  18                                                                                                               \n",
       "Pre_2 172                                                                                                               \n",
       "Pre_3   0                                                                                                               \n",
       "Pre_4  78                                                                                                               \n",
       "Pre_5  23                                                                                                               \n",
       "Pre_6 192                                                                                                               \n",
       "      k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_392\n",
       "Pre_1 15                                                                                                                \n",
       "Pre_2  0                                                                                                                \n",
       "Pre_3  0                                                                                                                \n",
       "Pre_4  0                                                                                                                \n",
       "Pre_5 33                                                                                                                \n",
       "Pre_6  0                                                                                                                \n",
       "      k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Leptotrichia.s__sp._HMT_212\n",
       "Pre_1  42                                                                                                               \n",
       "Pre_2  64                                                                                                               \n",
       "Pre_3   8                                                                                                               \n",
       "Pre_4  83                                                                                                               \n",
       "Pre_5 286                                                                                                               \n",
       "Pre_6   0                                                                                                               \n",
       "      k__Bacteria.p__Gracilibacteria_.GN02..c__Gracilibacteria_.GN02._.C.1..o__Gracilibacteria_.GN02._.O.1..f__Gracilibacteria_.GN02._.F.1..g__Gracilibacteria_.GN02._.G.1..s__bacterium_HMT_872\n",
       "Pre_1 0                                                                                                                                                                                         \n",
       "Pre_2 0                                                                                                                                                                                         \n",
       "Pre_3 0                                                                                                                                                                                         \n",
       "Pre_4 0                                                                                                                                                                                         \n",
       "Pre_5 0                                                                                                                                                                                         \n",
       "Pre_6 0                                                                                                                                                                                         \n",
       "      k__Bacteria.p__Firmicutes.c__Bacilli.o__Lactobacillales.f__Lactobacillaceae.g__Lacticaseibacillus.__\n",
       "Pre_1 0                                                                                                   \n",
       "Pre_2 0                                                                                                   \n",
       "Pre_3 0                                                                                                   \n",
       "Pre_4 0                                                                                                   \n",
       "Pre_5 0                                                                                                   \n",
       "Pre_6 0                                                                                                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# making the row names the sample ids\n",
    "baseline_saliva_df = saliva_df %>%\n",
    "    # only keep participants' baseline samples and those that didn't fail the IgE protein\n",
    "    filter(Time_Point == 'Pre' & IgE_Binarized != 'Fail') %>%\n",
    "    # ended up reassinging 'DS' to 'F' due to small sample size\n",
    "    mutate(Outcome = ifelse(Outcome == 'DS', 'F', Outcome)) %>%\n",
    "    # making the row names the sample ids\n",
    "    column_to_rownames(var = 'Sample_ID')\n",
    "\n",
    "baseline_stool_df = stool_df %>%\n",
    "    filter(Time_Point == 'Pre' & IgE_Binarized != 'Fail') %>%\n",
    "    mutate(Outcome = ifelse(Outcome == 'DS', 'F', Outcome)) %>%\n",
    "    column_to_rownames(var = 'Sample_ID')\n",
    "\n",
    "baseline_saliva_df$Outcome = factor(baseline_saliva_df$Outcome, levels = c('F', 'SU'))\n",
    "baseline_stool_df$Outcome = factor(baseline_stool_df$Outcome, levels = c('F', 'SU'))\n",
    "\n",
    "head(baseline_saliva_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "94f71add-206d-425a-912b-4257b08aa3f1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking the input data type ...\n",
      "\n",
      "The input data is of type: matrix\n",
      "\n",
      "The imported data is in a generic 'matrix'/'data.frame' format.\n",
      "\n",
      "PASS\n",
      "\n",
      "Checking the sample metadata ...\n",
      "\n",
      "The specified variables in the formula: Outcome\n",
      "\n",
      "The available variables in the sample metadata: Original_Sample_ID, Subject_No, Time_Point, Treatment, Outcome, Age, IgE, IgE_Binarized\n",
      "\n",
      "PASS\n",
      "\n",
      "Checking other arguments ...\n",
      "\n",
      "The number of groups of interest is: 2\n",
      "\n",
      "Warning message:\n",
      "“The group variable has < 3 categories \n",
      "The multi-group comparisons (global/pairwise/dunnet/trend) will be deactivated”\n",
      "The sample size per group is: F = 27, SU = 10\n",
      "\n",
      "PASS\n",
      "\n",
      "Obtaining initial estimates ...\n",
      "\n",
      "Estimating sample-specific biases ...\n",
      "\n",
      "Loading required package: foreach\n",
      "\n",
      "\n",
      "Attaching package: ‘foreach’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:purrr’:\n",
      "\n",
      "    accumulate, when\n",
      "\n",
      "\n",
      "Loading required package: rngtools\n",
      "\n",
      "Warning message in sqrt(sweep(var_hat, 2, var_delta, \"*\")):\n",
      "“NaNs produced”\n",
      "ANCOM-BC2 primary results ...\n",
      "\n",
      "Conducting sensitivity analysis for pseudo-count addition to 0s ...\n",
      "For taxa that are significant but do not pass the sensitivity analysis,\n",
      "they are marked in the 'passed_ss' column and will be treated as non-significant in the 'diff_robust' column.\n",
      "For detailed instructions on performing sensitivity analysis, please refer to the package vignette.\n",
      "\n",
      "Checking the input data type ...\n",
      "\n",
      "The input data is of type: matrix\n",
      "\n",
      "The imported data is in a generic 'matrix'/'data.frame' format.\n",
      "\n",
      "PASS\n",
      "\n",
      "Checking the sample metadata ...\n",
      "\n",
      "The specified variables in the formula: Age, IgE, Outcome\n",
      "\n",
      "The available variables in the sample metadata: Original_Sample_ID, Subject_No, Time_Point, Treatment, Outcome, Age, IgE, IgE_Binarized\n",
      "\n",
      "PASS\n",
      "\n",
      "Checking other arguments ...\n",
      "\n",
      "The number of groups of interest is: 2\n",
      "\n",
      "Warning message:\n",
      "“The group variable has < 3 categories \n",
      "The multi-group comparisons (global/pairwise/dunnet/trend) will be deactivated”\n",
      "The sample size per group is: F = 27, SU = 10\n",
      "\n",
      "PASS\n",
      "\n",
      "Obtaining initial estimates ...\n",
      "\n",
      "Estimating sample-specific biases ...\n",
      "\n",
      "ANCOM-BC2 primary results ...\n",
      "\n",
      "Warning message in pt(abs(W), df = dof, lower.tail = FALSE):\n",
      "“NaNs produced”\n",
      "Conducting sensitivity analysis for pseudo-count addition to 0s ...\n",
      "For taxa that are significant but do not pass the sensitivity analysis,\n",
      "they are marked in the 'passed_ss' column and will be treated as non-significant in the 'diff_robust' column.\n",
      "For detailed instructions on performing sensitivity analysis, please refer to the package vignette.\n",
      "\n",
      "Checking the input data type ...\n",
      "\n",
      "The input data is of type: matrix\n",
      "\n",
      "The imported data is in a generic 'matrix'/'data.frame' format.\n",
      "\n",
      "PASS\n",
      "\n",
      "Checking the sample metadata ...\n",
      "\n",
      "The specified variables in the formula: Outcome\n",
      "\n",
      "The available variables in the sample metadata: Original_Subject_ID, Original_Sample_ID, Subject_No, Time_Point, Treatment, Outcome, Age, IgE, IgE_Binarized\n",
      "\n",
      "PASS\n",
      "\n",
      "Checking other arguments ...\n",
      "\n",
      "The number of groups of interest is: 2\n",
      "\n",
      "Warning message:\n",
      "“The group variable has < 3 categories \n",
      "The multi-group comparisons (global/pairwise/dunnet/trend) will be deactivated”\n",
      "The sample size per group is: F = 24, SU = 10\n",
      "\n",
      "PASS\n",
      "\n",
      "Obtaining initial estimates ...\n",
      "\n",
      "Estimating sample-specific biases ...\n",
      "\n",
      "Warning message in sqrt(sweep(var_hat, 2, var_delta, \"*\")):\n",
      "“NaNs produced”\n",
      "ANCOM-BC2 primary results ...\n",
      "\n",
      "Conducting sensitivity analysis for pseudo-count addition to 0s ...\n",
      "For taxa that are significant but do not pass the sensitivity analysis,\n",
      "they are marked in the 'passed_ss' column and will be treated as non-significant in the 'diff_robust' column.\n",
      "For detailed instructions on performing sensitivity analysis, please refer to the package vignette.\n",
      "\n",
      "Checking the input data type ...\n",
      "\n",
      "The input data is of type: matrix\n",
      "\n",
      "The imported data is in a generic 'matrix'/'data.frame' format.\n",
      "\n",
      "PASS\n",
      "\n",
      "Checking the sample metadata ...\n",
      "\n",
      "The specified variables in the formula: Age, IgE, Outcome\n",
      "\n",
      "The available variables in the sample metadata: Original_Subject_ID, Original_Sample_ID, Subject_No, Time_Point, Treatment, Outcome, Age, IgE, IgE_Binarized\n",
      "\n",
      "PASS\n",
      "\n",
      "Checking other arguments ...\n",
      "\n",
      "The number of groups of interest is: 2\n",
      "\n",
      "Warning message:\n",
      "“The group variable has < 3 categories \n",
      "The multi-group comparisons (global/pairwise/dunnet/trend) will be deactivated”\n",
      "The sample size per group is: F = 24, SU = 10\n",
      "\n",
      "PASS\n",
      "\n",
      "Obtaining initial estimates ...\n",
      "\n",
      "Estimating sample-specific biases ...\n",
      "\n",
      "ANCOM-BC2 primary results ...\n",
      "\n",
      "Warning message in pt(abs(W), df = dof, lower.tail = FALSE):\n",
      "“NaNs produced”\n",
      "Conducting sensitivity analysis for pseudo-count addition to 0s ...\n",
      "For taxa that are significant but do not pass the sensitivity analysis,\n",
      "they are marked in the 'passed_ss' column and will be treated as non-significant in the 'diff_robust' column.\n",
      "For detailed instructions on performing sensitivity analysis, please refer to the package vignette.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# setting seed for reproducibility\n",
    "get_ancombc2_results = function(df, formula, value){\n",
    "    # add words!!!\n",
    "    \n",
    "    set.seed(123)\n",
    "    # automatically performs a psuedo log transformation\n",
    "    value2 = value + 1\n",
    "    output = ancombc2(data = t(df[,value2:ncol(df)]), \n",
    "                      meta_data = df[,1:value],\n",
    "                      tax_level = 'Species', #taxonomic level of interest\n",
    "                      fix_formula = formula, # specifying the model\n",
    "                      group = 'Outcome', # independent variable\n",
    "                      p_adj_method = \"holm\") # Holm-Bonferroni adj p values\n",
    "\n",
    "    # removing the standard error and test statistics from the results\n",
    "    removed_cols = c('se_', 'W_')\n",
    "    final_df = output$res %>%\n",
    "        select(!matches(removed_cols))\n",
    "\n",
    "    return(final_df)\n",
    "    }\n",
    "\n",
    "# calling fn\n",
    "saliva_raw_model = get_ancombc2_results(baseline_saliva_df, 'Outcome',8)\n",
    "saliva_2cov_model = get_ancombc2_results(baseline_saliva_df, 'Age + IgE + Outcome',8)\n",
    "stool_raw_model = get_ancombc2_results(baseline_stool_df, 'Outcome',9)\n",
    "stool_2cov_model = get_ancombc2_results(baseline_stool_df, 'Age + IgE + Outcome',9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb9ca100-5e7e-4618-b6fa-bae68180bf13",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "621ae9d7-3098-4116-b200-c43990e04a87",
   "metadata": {},
   "source": [
    "## Saliva Raw Results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e6ad0b5-d978-4987-9b00-d091ed12b544",
   "metadata": {},
   "source": [
    "## Saliva Adjusted Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ae6cac8e-9b4f-4e23-adee-c795b149a73d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>taxon</th><th scope=col>lfc_(Intercept)</th><th scope=col>lfc_OutcomeSU</th><th scope=col>p_(Intercept)</th><th scope=col>p_OutcomeSU</th><th scope=col>q_(Intercept)</th><th scope=col>q_OutcomeSU</th><th scope=col>diff_(Intercept)</th><th scope=col>diff_OutcomeSU</th><th scope=col>passed_ss_(Intercept)</th><th scope=col>passed_ss_OutcomeSU</th><th scope=col>diff_robust_(Intercept)</th><th scope=col>diff_robust_OutcomeSU</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " taxon & lfc\\_(Intercept) & lfc\\_OutcomeSU & p\\_(Intercept) & p\\_OutcomeSU & q\\_(Intercept) & q\\_OutcomeSU & diff\\_(Intercept) & diff\\_OutcomeSU & passed\\_ss\\_(Intercept) & passed\\_ss\\_OutcomeSU & diff\\_robust\\_(Intercept) & diff\\_robust\\_OutcomeSU\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 13\n",
       "\n",
       "| taxon &lt;chr&gt; | lfc_(Intercept) &lt;dbl&gt; | lfc_OutcomeSU &lt;dbl&gt; | p_(Intercept) &lt;dbl&gt; | p_OutcomeSU &lt;dbl&gt; | q_(Intercept) &lt;dbl&gt; | q_OutcomeSU &lt;dbl&gt; | diff_(Intercept) &lt;lgl&gt; | diff_OutcomeSU &lt;lgl&gt; | passed_ss_(Intercept) &lt;lgl&gt; | passed_ss_OutcomeSU &lt;lgl&gt; | diff_robust_(Intercept) &lt;lgl&gt; | diff_robust_OutcomeSU &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     taxon lfc_(Intercept) lfc_OutcomeSU p_(Intercept) p_OutcomeSU\n",
       "     q_(Intercept) q_OutcomeSU diff_(Intercept) diff_OutcomeSU\n",
       "     passed_ss_(Intercept) passed_ss_OutcomeSU diff_robust_(Intercept)\n",
       "     diff_robust_OutcomeSU"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 1 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>taxon</th><th scope=col>lfc_(Intercept)</th><th scope=col>lfc_OutcomeSU</th><th scope=col>p_(Intercept)</th><th scope=col>p_OutcomeSU</th><th scope=col>q_(Intercept)</th><th scope=col>q_OutcomeSU</th><th scope=col>diff_(Intercept)</th><th scope=col>diff_OutcomeSU</th><th scope=col>passed_ss_(Intercept)</th><th scope=col>passed_ss_OutcomeSU</th><th scope=col>diff_robust_(Intercept)</th><th scope=col>diff_robust_OutcomeSU</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Porphyromonadaceae.g__Porphyromonas.__</td><td>-0.5326396</td><td>1.611588</td><td>0.01091397</td><td>0.000414453</td><td>1</td><td>0.0609246</td><td>FALSE</td><td>FALSE</td><td>TRUE</td><td>TRUE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " taxon & lfc\\_(Intercept) & lfc\\_OutcomeSU & p\\_(Intercept) & p\\_OutcomeSU & q\\_(Intercept) & q\\_OutcomeSU & diff\\_(Intercept) & diff\\_OutcomeSU & passed\\_ss\\_(Intercept) & passed\\_ss\\_OutcomeSU & diff\\_robust\\_(Intercept) & diff\\_robust\\_OutcomeSU\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\t k\\_\\_Bacteria.p\\_\\_Bacteroidetes.c\\_\\_Bacteroidia.o\\_\\_Bacteroidales.f\\_\\_Porphyromonadaceae.g\\_\\_Porphyromonas.\\_\\_ & -0.5326396 & 1.611588 & 0.01091397 & 0.000414453 & 1 & 0.0609246 & FALSE & FALSE & TRUE & TRUE & FALSE & FALSE\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 13\n",
       "\n",
       "| taxon &lt;chr&gt; | lfc_(Intercept) &lt;dbl&gt; | lfc_OutcomeSU &lt;dbl&gt; | p_(Intercept) &lt;dbl&gt; | p_OutcomeSU &lt;dbl&gt; | q_(Intercept) &lt;dbl&gt; | q_OutcomeSU &lt;dbl&gt; | diff_(Intercept) &lt;lgl&gt; | diff_OutcomeSU &lt;lgl&gt; | passed_ss_(Intercept) &lt;lgl&gt; | passed_ss_OutcomeSU &lt;lgl&gt; | diff_robust_(Intercept) &lt;lgl&gt; | diff_robust_OutcomeSU &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Porphyromonadaceae.g__Porphyromonas.__ | -0.5326396 | 1.611588 | 0.01091397 | 0.000414453 | 1 | 0.0609246 | FALSE | FALSE | TRUE | TRUE | FALSE | FALSE |\n",
       "\n"
      ],
      "text/plain": [
       "  taxon                                                                                                 \n",
       "1 k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Porphyromonadaceae.g__Porphyromonas.__\n",
       "  lfc_(Intercept) lfc_OutcomeSU p_(Intercept) p_OutcomeSU q_(Intercept)\n",
       "1 -0.5326396      1.611588      0.01091397    0.000414453 1            \n",
       "  q_OutcomeSU diff_(Intercept) diff_OutcomeSU passed_ss_(Intercept)\n",
       "1 0.0609246   FALSE            FALSE          TRUE                 \n",
       "  passed_ss_OutcomeSU diff_robust_(Intercept) diff_robust_OutcomeSU\n",
       "1 TRUE                FALSE                   FALSE                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>taxon</th><th scope=col>lfc_(Intercept)</th><th scope=col>lfc_OutcomeSU</th><th scope=col>p_(Intercept)</th><th scope=col>p_OutcomeSU</th><th scope=col>q_(Intercept)</th><th scope=col>q_OutcomeSU</th><th scope=col>diff_(Intercept)</th><th scope=col>diff_OutcomeSU</th><th scope=col>passed_ss_(Intercept)</th><th scope=col>passed_ss_OutcomeSU</th><th scope=col>diff_robust_(Intercept)</th><th scope=col>diff_robust_OutcomeSU</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Prevotellaceae.g__Alloprevotella.s__sp._HMT_308         </td><td> 0.1434357</td><td>-2.310190</td><td>0.390295368</td><td>0.0001395744</td><td>1.0000000</td><td>0.02079658</td><td>FALSE</td><td> TRUE</td><td>TRUE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "\t<tr><td>k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Porphyromonadaceae.g__Porphyromonas.__                  </td><td>-0.5326396</td><td> 1.611588</td><td>0.010913973</td><td>0.0004144530</td><td>1.0000000</td><td>0.06092460</td><td>FALSE</td><td>FALSE</td><td>TRUE</td><td> TRUE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "\t<tr><td>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.s__sp._HMT_221</td><td>-0.7327041</td><td> 2.342115</td><td>0.001605976</td><td>0.0002212867</td><td>0.2264427</td><td>0.03275043</td><td>FALSE</td><td> TRUE</td><td>TRUE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " taxon & lfc\\_(Intercept) & lfc\\_OutcomeSU & p\\_(Intercept) & p\\_OutcomeSU & q\\_(Intercept) & q\\_OutcomeSU & diff\\_(Intercept) & diff\\_OutcomeSU & passed\\_ss\\_(Intercept) & passed\\_ss\\_OutcomeSU & diff\\_robust\\_(Intercept) & diff\\_robust\\_OutcomeSU\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\t k\\_\\_Bacteria.p\\_\\_Bacteroidetes.c\\_\\_Bacteroidia.o\\_\\_Bacteroidales.f\\_\\_Prevotellaceae.g\\_\\_Alloprevotella.s\\_\\_sp.\\_HMT\\_308          &  0.1434357 & -2.310190 & 0.390295368 & 0.0001395744 & 1.0000000 & 0.02079658 & FALSE &  TRUE & TRUE & FALSE & FALSE & FALSE\\\\\n",
       "\t k\\_\\_Bacteria.p\\_\\_Bacteroidetes.c\\_\\_Bacteroidia.o\\_\\_Bacteroidales.f\\_\\_Porphyromonadaceae.g\\_\\_Porphyromonas.\\_\\_                   & -0.5326396 &  1.611588 & 0.010913973 & 0.0004144530 & 1.0000000 & 0.06092460 & FALSE & FALSE & TRUE &  TRUE & FALSE & FALSE\\\\\n",
       "\t k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Pseudoleptotrichia.s\\_\\_sp.\\_HMT\\_221 & -0.7327041 &  2.342115 & 0.001605976 & 0.0002212867 & 0.2264427 & 0.03275043 & FALSE &  TRUE & TRUE & FALSE & FALSE & FALSE\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 13\n",
       "\n",
       "| taxon &lt;chr&gt; | lfc_(Intercept) &lt;dbl&gt; | lfc_OutcomeSU &lt;dbl&gt; | p_(Intercept) &lt;dbl&gt; | p_OutcomeSU &lt;dbl&gt; | q_(Intercept) &lt;dbl&gt; | q_OutcomeSU &lt;dbl&gt; | diff_(Intercept) &lt;lgl&gt; | diff_OutcomeSU &lt;lgl&gt; | passed_ss_(Intercept) &lt;lgl&gt; | passed_ss_OutcomeSU &lt;lgl&gt; | diff_robust_(Intercept) &lt;lgl&gt; | diff_robust_OutcomeSU &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Prevotellaceae.g__Alloprevotella.s__sp._HMT_308          |  0.1434357 | -2.310190 | 0.390295368 | 0.0001395744 | 1.0000000 | 0.02079658 | FALSE |  TRUE | TRUE | FALSE | FALSE | FALSE |\n",
       "| k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Porphyromonadaceae.g__Porphyromonas.__                   | -0.5326396 |  1.611588 | 0.010913973 | 0.0004144530 | 1.0000000 | 0.06092460 | FALSE | FALSE | TRUE |  TRUE | FALSE | FALSE |\n",
       "| k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.s__sp._HMT_221 | -0.7327041 |  2.342115 | 0.001605976 | 0.0002212867 | 0.2264427 | 0.03275043 | FALSE |  TRUE | TRUE | FALSE | FALSE | FALSE |\n",
       "\n"
      ],
      "text/plain": [
       "  taxon                                                                                                                   \n",
       "1 k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Prevotellaceae.g__Alloprevotella.s__sp._HMT_308         \n",
       "2 k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Porphyromonadaceae.g__Porphyromonas.__                  \n",
       "3 k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.s__sp._HMT_221\n",
       "  lfc_(Intercept) lfc_OutcomeSU p_(Intercept) p_OutcomeSU  q_(Intercept)\n",
       "1  0.1434357      -2.310190     0.390295368   0.0001395744 1.0000000    \n",
       "2 -0.5326396       1.611588     0.010913973   0.0004144530 1.0000000    \n",
       "3 -0.7327041       2.342115     0.001605976   0.0002212867 0.2264427    \n",
       "  q_OutcomeSU diff_(Intercept) diff_OutcomeSU passed_ss_(Intercept)\n",
       "1 0.02079658  FALSE             TRUE          TRUE                 \n",
       "2 0.06092460  FALSE            FALSE          TRUE                 \n",
       "3 0.03275043  FALSE             TRUE          TRUE                 \n",
       "  passed_ss_OutcomeSU diff_robust_(Intercept) diff_robust_OutcomeSU\n",
       "1 FALSE               FALSE                   FALSE                \n",
       "2  TRUE               FALSE                   FALSE                \n",
       "3 FALSE               FALSE                   FALSE                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# passed sensitivity filter and is significant \n",
    "saliva_raw_model %>%\n",
    "    filter(diff_robust_OutcomeSU == TRUE)\n",
    "\n",
    "# passed sensitivity analysis with a relaxed filter\n",
    "saliva_raw_model %>%\n",
    "    filter(q_OutcomeSU < 0.1 & passed_ss_OutcomeSU == TRUE)\n",
    "\n",
    "# significant with a relaxed filter\n",
    "saliva_raw_model %>%\n",
    "    filter(q_OutcomeSU < 0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31fd222a-0fba-45cd-ab8f-87d1d676910b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01f15a02-28b1-4a49-a490-6f130a134770",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2800505a-2275-4db8-ab4e-e89fbb887e23",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 25</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>taxon</th><th scope=col>lfc_(Intercept)</th><th scope=col>lfc_Age</th><th scope=col>lfc_IgE</th><th scope=col>lfc_OutcomeSU</th><th scope=col>p_(Intercept)</th><th scope=col>p_Age</th><th scope=col>p_IgE</th><th scope=col>p_OutcomeSU</th><th scope=col>q_(Intercept)</th><th scope=col>⋯</th><th scope=col>diff_IgE</th><th scope=col>diff_OutcomeSU</th><th scope=col>passed_ss_(Intercept)</th><th scope=col>passed_ss_Age</th><th scope=col>passed_ss_IgE</th><th scope=col>passed_ss_OutcomeSU</th><th scope=col>diff_robust_(Intercept)</th><th scope=col>diff_robust_Age</th><th scope=col>diff_robust_IgE</th><th scope=col>diff_robust_OutcomeSU</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 25\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " taxon & lfc\\_(Intercept) & lfc\\_Age & lfc\\_IgE & lfc\\_OutcomeSU & p\\_(Intercept) & p\\_Age & p\\_IgE & p\\_OutcomeSU & q\\_(Intercept) & ⋯ & diff\\_IgE & diff\\_OutcomeSU & passed\\_ss\\_(Intercept) & passed\\_ss\\_Age & passed\\_ss\\_IgE & passed\\_ss\\_OutcomeSU & diff\\_robust\\_(Intercept) & diff\\_robust\\_Age & diff\\_robust\\_IgE & diff\\_robust\\_OutcomeSU\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 25\n",
       "\n",
       "| taxon &lt;chr&gt; | lfc_(Intercept) &lt;dbl&gt; | lfc_Age &lt;dbl&gt; | lfc_IgE &lt;dbl&gt; | lfc_OutcomeSU &lt;dbl&gt; | p_(Intercept) &lt;dbl&gt; | p_Age &lt;dbl&gt; | p_IgE &lt;dbl&gt; | p_OutcomeSU &lt;dbl&gt; | q_(Intercept) &lt;dbl&gt; | ⋯ ⋯ | diff_IgE &lt;lgl&gt; | diff_OutcomeSU &lt;lgl&gt; | passed_ss_(Intercept) &lt;lgl&gt; | passed_ss_Age &lt;lgl&gt; | passed_ss_IgE &lt;lgl&gt; | passed_ss_OutcomeSU &lt;lgl&gt; | diff_robust_(Intercept) &lt;lgl&gt; | diff_robust_Age &lt;lgl&gt; | diff_robust_IgE &lt;lgl&gt; | diff_robust_OutcomeSU &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     taxon lfc_(Intercept) lfc_Age lfc_IgE lfc_OutcomeSU p_(Intercept) p_Age\n",
       "     p_IgE p_OutcomeSU q_(Intercept) ⋯ diff_IgE diff_OutcomeSU\n",
       "     passed_ss_(Intercept) passed_ss_Age passed_ss_IgE passed_ss_OutcomeSU\n",
       "     diff_robust_(Intercept) diff_robust_Age diff_robust_IgE\n",
       "     diff_robust_OutcomeSU"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 25</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>taxon</th><th scope=col>lfc_(Intercept)</th><th scope=col>lfc_Age</th><th scope=col>lfc_IgE</th><th scope=col>lfc_OutcomeSU</th><th scope=col>p_(Intercept)</th><th scope=col>p_Age</th><th scope=col>p_IgE</th><th scope=col>p_OutcomeSU</th><th scope=col>q_(Intercept)</th><th scope=col>⋯</th><th scope=col>diff_IgE</th><th scope=col>diff_OutcomeSU</th><th scope=col>passed_ss_(Intercept)</th><th scope=col>passed_ss_Age</th><th scope=col>passed_ss_IgE</th><th scope=col>passed_ss_OutcomeSU</th><th scope=col>diff_robust_(Intercept)</th><th scope=col>diff_robust_Age</th><th scope=col>diff_robust_IgE</th><th scope=col>diff_robust_OutcomeSU</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 25\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " taxon & lfc\\_(Intercept) & lfc\\_Age & lfc\\_IgE & lfc\\_OutcomeSU & p\\_(Intercept) & p\\_Age & p\\_IgE & p\\_OutcomeSU & q\\_(Intercept) & ⋯ & diff\\_IgE & diff\\_OutcomeSU & passed\\_ss\\_(Intercept) & passed\\_ss\\_Age & passed\\_ss\\_IgE & passed\\_ss\\_OutcomeSU & diff\\_robust\\_(Intercept) & diff\\_robust\\_Age & diff\\_robust\\_IgE & diff\\_robust\\_OutcomeSU\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 25\n",
       "\n",
       "| taxon &lt;chr&gt; | lfc_(Intercept) &lt;dbl&gt; | lfc_Age &lt;dbl&gt; | lfc_IgE &lt;dbl&gt; | lfc_OutcomeSU &lt;dbl&gt; | p_(Intercept) &lt;dbl&gt; | p_Age &lt;dbl&gt; | p_IgE &lt;dbl&gt; | p_OutcomeSU &lt;dbl&gt; | q_(Intercept) &lt;dbl&gt; | ⋯ ⋯ | diff_IgE &lt;lgl&gt; | diff_OutcomeSU &lt;lgl&gt; | passed_ss_(Intercept) &lt;lgl&gt; | passed_ss_Age &lt;lgl&gt; | passed_ss_IgE &lt;lgl&gt; | passed_ss_OutcomeSU &lt;lgl&gt; | diff_robust_(Intercept) &lt;lgl&gt; | diff_robust_Age &lt;lgl&gt; | diff_robust_IgE &lt;lgl&gt; | diff_robust_OutcomeSU &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     taxon lfc_(Intercept) lfc_Age lfc_IgE lfc_OutcomeSU p_(Intercept) p_Age\n",
       "     p_IgE p_OutcomeSU q_(Intercept) ⋯ diff_IgE diff_OutcomeSU\n",
       "     passed_ss_(Intercept) passed_ss_Age passed_ss_IgE passed_ss_OutcomeSU\n",
       "     diff_robust_(Intercept) diff_robust_Age diff_robust_IgE\n",
       "     diff_robust_OutcomeSU"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 25</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>taxon</th><th scope=col>lfc_(Intercept)</th><th scope=col>lfc_Age</th><th scope=col>lfc_IgE</th><th scope=col>lfc_OutcomeSU</th><th scope=col>p_(Intercept)</th><th scope=col>p_Age</th><th scope=col>p_IgE</th><th scope=col>p_OutcomeSU</th><th scope=col>q_(Intercept)</th><th scope=col>⋯</th><th scope=col>diff_IgE</th><th scope=col>diff_OutcomeSU</th><th scope=col>passed_ss_(Intercept)</th><th scope=col>passed_ss_Age</th><th scope=col>passed_ss_IgE</th><th scope=col>passed_ss_OutcomeSU</th><th scope=col>diff_robust_(Intercept)</th><th scope=col>diff_robust_Age</th><th scope=col>diff_robust_IgE</th><th scope=col>diff_robust_OutcomeSU</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 25\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " taxon & lfc\\_(Intercept) & lfc\\_Age & lfc\\_IgE & lfc\\_OutcomeSU & p\\_(Intercept) & p\\_Age & p\\_IgE & p\\_OutcomeSU & q\\_(Intercept) & ⋯ & diff\\_IgE & diff\\_OutcomeSU & passed\\_ss\\_(Intercept) & passed\\_ss\\_Age & passed\\_ss\\_IgE & passed\\_ss\\_OutcomeSU & diff\\_robust\\_(Intercept) & diff\\_robust\\_Age & diff\\_robust\\_IgE & diff\\_robust\\_OutcomeSU\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 25\n",
       "\n",
       "| taxon &lt;chr&gt; | lfc_(Intercept) &lt;dbl&gt; | lfc_Age &lt;dbl&gt; | lfc_IgE &lt;dbl&gt; | lfc_OutcomeSU &lt;dbl&gt; | p_(Intercept) &lt;dbl&gt; | p_Age &lt;dbl&gt; | p_IgE &lt;dbl&gt; | p_OutcomeSU &lt;dbl&gt; | q_(Intercept) &lt;dbl&gt; | ⋯ ⋯ | diff_IgE &lt;lgl&gt; | diff_OutcomeSU &lt;lgl&gt; | passed_ss_(Intercept) &lt;lgl&gt; | passed_ss_Age &lt;lgl&gt; | passed_ss_IgE &lt;lgl&gt; | passed_ss_OutcomeSU &lt;lgl&gt; | diff_robust_(Intercept) &lt;lgl&gt; | diff_robust_Age &lt;lgl&gt; | diff_robust_IgE &lt;lgl&gt; | diff_robust_OutcomeSU &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     taxon lfc_(Intercept) lfc_Age lfc_IgE lfc_OutcomeSU p_(Intercept) p_Age\n",
       "     p_IgE p_OutcomeSU q_(Intercept) ⋯ diff_IgE diff_OutcomeSU\n",
       "     passed_ss_(Intercept) passed_ss_Age passed_ss_IgE passed_ss_OutcomeSU\n",
       "     diff_robust_(Intercept) diff_robust_Age diff_robust_IgE\n",
       "     diff_robust_OutcomeSU"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# passed sensitivity filter and is significant \n",
    "saliva_2cov_model %>%\n",
    "    filter(diff_robust_OutcomeSU == TRUE)\n",
    "\n",
    "# passed sensitivity analysis with a relaxed filter\n",
    "saliva_2cov_model %>%\n",
    "    filter(q_OutcomeSU < 0.1 & passed_ss_OutcomeSU == TRUE)\n",
    "\n",
    "# significant with a relaxed filter\n",
    "saliva_2cov_model %>%\n",
    "    filter(q_OutcomeSU < 0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6b226af-e3ab-45f5-8949-4111f4247fd3",
   "metadata": {},
   "source": [
    "## Stool Raw Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "16a7f9ab-4cad-4307-b96c-5fe061a924c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>taxon</th><th scope=col>lfc_(Intercept)</th><th scope=col>lfc_OutcomeSU</th><th scope=col>p_(Intercept)</th><th scope=col>p_OutcomeSU</th><th scope=col>q_(Intercept)</th><th scope=col>q_OutcomeSU</th><th scope=col>diff_(Intercept)</th><th scope=col>diff_OutcomeSU</th><th scope=col>passed_ss_(Intercept)</th><th scope=col>passed_ss_OutcomeSU</th><th scope=col>diff_robust_(Intercept)</th><th scope=col>diff_robust_OutcomeSU</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " taxon & lfc\\_(Intercept) & lfc\\_OutcomeSU & p\\_(Intercept) & p\\_OutcomeSU & q\\_(Intercept) & q\\_OutcomeSU & diff\\_(Intercept) & diff\\_OutcomeSU & passed\\_ss\\_(Intercept) & passed\\_ss\\_OutcomeSU & diff\\_robust\\_(Intercept) & diff\\_robust\\_OutcomeSU\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 13\n",
       "\n",
       "| taxon &lt;chr&gt; | lfc_(Intercept) &lt;dbl&gt; | lfc_OutcomeSU &lt;dbl&gt; | p_(Intercept) &lt;dbl&gt; | p_OutcomeSU &lt;dbl&gt; | q_(Intercept) &lt;dbl&gt; | q_OutcomeSU &lt;dbl&gt; | diff_(Intercept) &lt;lgl&gt; | diff_OutcomeSU &lt;lgl&gt; | passed_ss_(Intercept) &lt;lgl&gt; | passed_ss_OutcomeSU &lt;lgl&gt; | diff_robust_(Intercept) &lt;lgl&gt; | diff_robust_OutcomeSU &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     taxon lfc_(Intercept) lfc_OutcomeSU p_(Intercept) p_OutcomeSU\n",
       "     q_(Intercept) q_OutcomeSU diff_(Intercept) diff_OutcomeSU\n",
       "     passed_ss_(Intercept) passed_ss_OutcomeSU diff_robust_(Intercept)\n",
       "     diff_robust_OutcomeSU"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 1 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>taxon</th><th scope=col>lfc_(Intercept)</th><th scope=col>lfc_OutcomeSU</th><th scope=col>p_(Intercept)</th><th scope=col>p_OutcomeSU</th><th scope=col>q_(Intercept)</th><th scope=col>q_OutcomeSU</th><th scope=col>diff_(Intercept)</th><th scope=col>diff_OutcomeSU</th><th scope=col>passed_ss_(Intercept)</th><th scope=col>passed_ss_OutcomeSU</th><th scope=col>diff_robust_(Intercept)</th><th scope=col>diff_robust_OutcomeSU</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>d__Bacteria.p__Actinobacteriota.c__Actinomycetia.o__Actinomycetales.f__Bifidobacteriaceae.g__Bifidobacterium_388775.s__Bifidobacterium.bifidum</td><td>0.8605398</td><td>-2.814578</td><td>0.000276249</td><td>0.0003467552</td><td>0.06574727</td><td>0.08322125</td><td>FALSE</td><td>FALSE</td><td>TRUE</td><td>TRUE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " taxon & lfc\\_(Intercept) & lfc\\_OutcomeSU & p\\_(Intercept) & p\\_OutcomeSU & q\\_(Intercept) & q\\_OutcomeSU & diff\\_(Intercept) & diff\\_OutcomeSU & passed\\_ss\\_(Intercept) & passed\\_ss\\_OutcomeSU & diff\\_robust\\_(Intercept) & diff\\_robust\\_OutcomeSU\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\t d\\_\\_Bacteria.p\\_\\_Actinobacteriota.c\\_\\_Actinomycetia.o\\_\\_Actinomycetales.f\\_\\_Bifidobacteriaceae.g\\_\\_Bifidobacterium\\_388775.s\\_\\_Bifidobacterium.bifidum & 0.8605398 & -2.814578 & 0.000276249 & 0.0003467552 & 0.06574727 & 0.08322125 & FALSE & FALSE & TRUE & TRUE & FALSE & FALSE\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 13\n",
       "\n",
       "| taxon &lt;chr&gt; | lfc_(Intercept) &lt;dbl&gt; | lfc_OutcomeSU &lt;dbl&gt; | p_(Intercept) &lt;dbl&gt; | p_OutcomeSU &lt;dbl&gt; | q_(Intercept) &lt;dbl&gt; | q_OutcomeSU &lt;dbl&gt; | diff_(Intercept) &lt;lgl&gt; | diff_OutcomeSU &lt;lgl&gt; | passed_ss_(Intercept) &lt;lgl&gt; | passed_ss_OutcomeSU &lt;lgl&gt; | diff_robust_(Intercept) &lt;lgl&gt; | diff_robust_OutcomeSU &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| d__Bacteria.p__Actinobacteriota.c__Actinomycetia.o__Actinomycetales.f__Bifidobacteriaceae.g__Bifidobacterium_388775.s__Bifidobacterium.bifidum | 0.8605398 | -2.814578 | 0.000276249 | 0.0003467552 | 0.06574727 | 0.08322125 | FALSE | FALSE | TRUE | TRUE | FALSE | FALSE |\n",
       "\n"
      ],
      "text/plain": [
       "  taxon                                                                                                                                         \n",
       "1 d__Bacteria.p__Actinobacteriota.c__Actinomycetia.o__Actinomycetales.f__Bifidobacteriaceae.g__Bifidobacterium_388775.s__Bifidobacterium.bifidum\n",
       "  lfc_(Intercept) lfc_OutcomeSU p_(Intercept) p_OutcomeSU  q_(Intercept)\n",
       "1 0.8605398       -2.814578     0.000276249   0.0003467552 0.06574727   \n",
       "  q_OutcomeSU diff_(Intercept) diff_OutcomeSU passed_ss_(Intercept)\n",
       "1 0.08322125  FALSE            FALSE          TRUE                 \n",
       "  passed_ss_OutcomeSU diff_robust_(Intercept) diff_robust_OutcomeSU\n",
       "1 TRUE                FALSE                   FALSE                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>taxon</th><th scope=col>lfc_(Intercept)</th><th scope=col>lfc_OutcomeSU</th><th scope=col>p_(Intercept)</th><th scope=col>p_OutcomeSU</th><th scope=col>q_(Intercept)</th><th scope=col>q_OutcomeSU</th><th scope=col>diff_(Intercept)</th><th scope=col>diff_OutcomeSU</th><th scope=col>passed_ss_(Intercept)</th><th scope=col>passed_ss_OutcomeSU</th><th scope=col>diff_robust_(Intercept)</th><th scope=col>diff_robust_OutcomeSU</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>d__Bacteria.p__Firmicutes_A.c__Clostridia_258483.o__Oscillospirales.f__Ruminococcaceae.g__Gemmiger_A_73129.s__Gemmiger_A_73129.qucibialis     </td><td>-1.2676072</td><td> 4.376294</td><td>0.003226683</td><td>1.180502e-04</td><td>0.74859043</td><td>0.02845009</td><td>FALSE</td><td> TRUE</td><td>TRUE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "\t<tr><td>d__Bacteria.p__Actinobacteriota.c__Actinomycetia.o__Actinomycetales.f__Bifidobacteriaceae.g__Bifidobacterium_388775.s__Bifidobacterium.bifidum</td><td> 0.8605398</td><td>-2.814578</td><td>0.000276249</td><td>3.467552e-04</td><td>0.06574727</td><td>0.08322125</td><td>FALSE</td><td>FALSE</td><td>TRUE</td><td> TRUE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "\t<tr><td>d__Bacteria.p__Firmicutes_A.c__Clostridia_258483.o__Peptostreptococcales.f__Peptostreptococcaceae_256921.g__Terrisporobacter.__               </td><td>-0.7070190</td><td> 3.118313</td><td>0.061454756</td><td>8.179198e-05</td><td>1.00000000</td><td>0.01979366</td><td>FALSE</td><td> TRUE</td><td>TRUE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " taxon & lfc\\_(Intercept) & lfc\\_OutcomeSU & p\\_(Intercept) & p\\_OutcomeSU & q\\_(Intercept) & q\\_OutcomeSU & diff\\_(Intercept) & diff\\_OutcomeSU & passed\\_ss\\_(Intercept) & passed\\_ss\\_OutcomeSU & diff\\_robust\\_(Intercept) & diff\\_robust\\_OutcomeSU\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\t d\\_\\_Bacteria.p\\_\\_Firmicutes\\_A.c\\_\\_Clostridia\\_258483.o\\_\\_Oscillospirales.f\\_\\_Ruminococcaceae.g\\_\\_Gemmiger\\_A\\_73129.s\\_\\_Gemmiger\\_A\\_73129.qucibialis      & -1.2676072 &  4.376294 & 0.003226683 & 1.180502e-04 & 0.74859043 & 0.02845009 & FALSE &  TRUE & TRUE & FALSE & FALSE & FALSE\\\\\n",
       "\t d\\_\\_Bacteria.p\\_\\_Actinobacteriota.c\\_\\_Actinomycetia.o\\_\\_Actinomycetales.f\\_\\_Bifidobacteriaceae.g\\_\\_Bifidobacterium\\_388775.s\\_\\_Bifidobacterium.bifidum &  0.8605398 & -2.814578 & 0.000276249 & 3.467552e-04 & 0.06574727 & 0.08322125 & FALSE & FALSE & TRUE &  TRUE & FALSE & FALSE\\\\\n",
       "\t d\\_\\_Bacteria.p\\_\\_Firmicutes\\_A.c\\_\\_Clostridia\\_258483.o\\_\\_Peptostreptococcales.f\\_\\_Peptostreptococcaceae\\_256921.g\\_\\_Terrisporobacter.\\_\\_                & -0.7070190 &  3.118313 & 0.061454756 & 8.179198e-05 & 1.00000000 & 0.01979366 & FALSE &  TRUE & TRUE & FALSE & FALSE & FALSE\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 13\n",
       "\n",
       "| taxon &lt;chr&gt; | lfc_(Intercept) &lt;dbl&gt; | lfc_OutcomeSU &lt;dbl&gt; | p_(Intercept) &lt;dbl&gt; | p_OutcomeSU &lt;dbl&gt; | q_(Intercept) &lt;dbl&gt; | q_OutcomeSU &lt;dbl&gt; | diff_(Intercept) &lt;lgl&gt; | diff_OutcomeSU &lt;lgl&gt; | passed_ss_(Intercept) &lt;lgl&gt; | passed_ss_OutcomeSU &lt;lgl&gt; | diff_robust_(Intercept) &lt;lgl&gt; | diff_robust_OutcomeSU &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| d__Bacteria.p__Firmicutes_A.c__Clostridia_258483.o__Oscillospirales.f__Ruminococcaceae.g__Gemmiger_A_73129.s__Gemmiger_A_73129.qucibialis      | -1.2676072 |  4.376294 | 0.003226683 | 1.180502e-04 | 0.74859043 | 0.02845009 | FALSE |  TRUE | TRUE | FALSE | FALSE | FALSE |\n",
       "| d__Bacteria.p__Actinobacteriota.c__Actinomycetia.o__Actinomycetales.f__Bifidobacteriaceae.g__Bifidobacterium_388775.s__Bifidobacterium.bifidum |  0.8605398 | -2.814578 | 0.000276249 | 3.467552e-04 | 0.06574727 | 0.08322125 | FALSE | FALSE | TRUE |  TRUE | FALSE | FALSE |\n",
       "| d__Bacteria.p__Firmicutes_A.c__Clostridia_258483.o__Peptostreptococcales.f__Peptostreptococcaceae_256921.g__Terrisporobacter.__                | -0.7070190 |  3.118313 | 0.061454756 | 8.179198e-05 | 1.00000000 | 0.01979366 | FALSE |  TRUE | TRUE | FALSE | FALSE | FALSE |\n",
       "\n"
      ],
      "text/plain": [
       "  taxon                                                                                                                                         \n",
       "1 d__Bacteria.p__Firmicutes_A.c__Clostridia_258483.o__Oscillospirales.f__Ruminococcaceae.g__Gemmiger_A_73129.s__Gemmiger_A_73129.qucibialis     \n",
       "2 d__Bacteria.p__Actinobacteriota.c__Actinomycetia.o__Actinomycetales.f__Bifidobacteriaceae.g__Bifidobacterium_388775.s__Bifidobacterium.bifidum\n",
       "3 d__Bacteria.p__Firmicutes_A.c__Clostridia_258483.o__Peptostreptococcales.f__Peptostreptococcaceae_256921.g__Terrisporobacter.__               \n",
       "  lfc_(Intercept) lfc_OutcomeSU p_(Intercept) p_OutcomeSU  q_(Intercept)\n",
       "1 -1.2676072       4.376294     0.003226683   1.180502e-04 0.74859043   \n",
       "2  0.8605398      -2.814578     0.000276249   3.467552e-04 0.06574727   \n",
       "3 -0.7070190       3.118313     0.061454756   8.179198e-05 1.00000000   \n",
       "  q_OutcomeSU diff_(Intercept) diff_OutcomeSU passed_ss_(Intercept)\n",
       "1 0.02845009  FALSE             TRUE          TRUE                 \n",
       "2 0.08322125  FALSE            FALSE          TRUE                 \n",
       "3 0.01979366  FALSE             TRUE          TRUE                 \n",
       "  passed_ss_OutcomeSU diff_robust_(Intercept) diff_robust_OutcomeSU\n",
       "1 FALSE               FALSE                   FALSE                \n",
       "2  TRUE               FALSE                   FALSE                \n",
       "3 FALSE               FALSE                   FALSE                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# passed sensitivity filter and is significant \n",
    "stool_raw_model %>%\n",
    "    filter(diff_robust_OutcomeSU == TRUE)\n",
    "\n",
    "# passed sensitivity analysis with a relaxed filter\n",
    "stool_raw_model %>%\n",
    "    filter(q_OutcomeSU < 0.1 & passed_ss_OutcomeSU == TRUE)\n",
    "\n",
    "# significant with a relaxed filter\n",
    "stool_raw_model %>%\n",
    "    filter(q_OutcomeSU < 0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f145b1b-75e3-48b6-a965-7e094c49fe68",
   "metadata": {},
   "source": [
    "## Stool Adjusted Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "47a7cea2-1fa7-4776-b94d-4806113dd7e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 25</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>taxon</th><th scope=col>lfc_(Intercept)</th><th scope=col>lfc_Age</th><th scope=col>lfc_IgE</th><th scope=col>lfc_OutcomeSU</th><th scope=col>p_(Intercept)</th><th scope=col>p_Age</th><th scope=col>p_IgE</th><th scope=col>p_OutcomeSU</th><th scope=col>q_(Intercept)</th><th scope=col>⋯</th><th scope=col>diff_IgE</th><th scope=col>diff_OutcomeSU</th><th scope=col>passed_ss_(Intercept)</th><th scope=col>passed_ss_Age</th><th scope=col>passed_ss_IgE</th><th scope=col>passed_ss_OutcomeSU</th><th scope=col>diff_robust_(Intercept)</th><th scope=col>diff_robust_Age</th><th scope=col>diff_robust_IgE</th><th scope=col>diff_robust_OutcomeSU</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 25\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " taxon & lfc\\_(Intercept) & lfc\\_Age & lfc\\_IgE & lfc\\_OutcomeSU & p\\_(Intercept) & p\\_Age & p\\_IgE & p\\_OutcomeSU & q\\_(Intercept) & ⋯ & diff\\_IgE & diff\\_OutcomeSU & passed\\_ss\\_(Intercept) & passed\\_ss\\_Age & passed\\_ss\\_IgE & passed\\_ss\\_OutcomeSU & diff\\_robust\\_(Intercept) & diff\\_robust\\_Age & diff\\_robust\\_IgE & diff\\_robust\\_OutcomeSU\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 25\n",
       "\n",
       "| taxon &lt;chr&gt; | lfc_(Intercept) &lt;dbl&gt; | lfc_Age &lt;dbl&gt; | lfc_IgE &lt;dbl&gt; | lfc_OutcomeSU &lt;dbl&gt; | p_(Intercept) &lt;dbl&gt; | p_Age &lt;dbl&gt; | p_IgE &lt;dbl&gt; | p_OutcomeSU &lt;dbl&gt; | q_(Intercept) &lt;dbl&gt; | ⋯ ⋯ | diff_IgE &lt;lgl&gt; | diff_OutcomeSU &lt;lgl&gt; | passed_ss_(Intercept) &lt;lgl&gt; | passed_ss_Age &lt;lgl&gt; | passed_ss_IgE &lt;lgl&gt; | passed_ss_OutcomeSU &lt;lgl&gt; | diff_robust_(Intercept) &lt;lgl&gt; | diff_robust_Age &lt;lgl&gt; | diff_robust_IgE &lt;lgl&gt; | diff_robust_OutcomeSU &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     taxon lfc_(Intercept) lfc_Age lfc_IgE lfc_OutcomeSU p_(Intercept) p_Age\n",
       "     p_IgE p_OutcomeSU q_(Intercept) ⋯ diff_IgE diff_OutcomeSU\n",
       "     passed_ss_(Intercept) passed_ss_Age passed_ss_IgE passed_ss_OutcomeSU\n",
       "     diff_robust_(Intercept) diff_robust_Age diff_robust_IgE\n",
       "     diff_robust_OutcomeSU"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 25</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>taxon</th><th scope=col>lfc_(Intercept)</th><th scope=col>lfc_Age</th><th scope=col>lfc_IgE</th><th scope=col>lfc_OutcomeSU</th><th scope=col>p_(Intercept)</th><th scope=col>p_Age</th><th scope=col>p_IgE</th><th scope=col>p_OutcomeSU</th><th scope=col>q_(Intercept)</th><th scope=col>⋯</th><th scope=col>diff_IgE</th><th scope=col>diff_OutcomeSU</th><th scope=col>passed_ss_(Intercept)</th><th scope=col>passed_ss_Age</th><th scope=col>passed_ss_IgE</th><th scope=col>passed_ss_OutcomeSU</th><th scope=col>diff_robust_(Intercept)</th><th scope=col>diff_robust_Age</th><th scope=col>diff_robust_IgE</th><th scope=col>diff_robust_OutcomeSU</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 25\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " taxon & lfc\\_(Intercept) & lfc\\_Age & lfc\\_IgE & lfc\\_OutcomeSU & p\\_(Intercept) & p\\_Age & p\\_IgE & p\\_OutcomeSU & q\\_(Intercept) & ⋯ & diff\\_IgE & diff\\_OutcomeSU & passed\\_ss\\_(Intercept) & passed\\_ss\\_Age & passed\\_ss\\_IgE & passed\\_ss\\_OutcomeSU & diff\\_robust\\_(Intercept) & diff\\_robust\\_Age & diff\\_robust\\_IgE & diff\\_robust\\_OutcomeSU\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 25\n",
       "\n",
       "| taxon &lt;chr&gt; | lfc_(Intercept) &lt;dbl&gt; | lfc_Age &lt;dbl&gt; | lfc_IgE &lt;dbl&gt; | lfc_OutcomeSU &lt;dbl&gt; | p_(Intercept) &lt;dbl&gt; | p_Age &lt;dbl&gt; | p_IgE &lt;dbl&gt; | p_OutcomeSU &lt;dbl&gt; | q_(Intercept) &lt;dbl&gt; | ⋯ ⋯ | diff_IgE &lt;lgl&gt; | diff_OutcomeSU &lt;lgl&gt; | passed_ss_(Intercept) &lt;lgl&gt; | passed_ss_Age &lt;lgl&gt; | passed_ss_IgE &lt;lgl&gt; | passed_ss_OutcomeSU &lt;lgl&gt; | diff_robust_(Intercept) &lt;lgl&gt; | diff_robust_Age &lt;lgl&gt; | diff_robust_IgE &lt;lgl&gt; | diff_robust_OutcomeSU &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     taxon lfc_(Intercept) lfc_Age lfc_IgE lfc_OutcomeSU p_(Intercept) p_Age\n",
       "     p_IgE p_OutcomeSU q_(Intercept) ⋯ diff_IgE diff_OutcomeSU\n",
       "     passed_ss_(Intercept) passed_ss_Age passed_ss_IgE passed_ss_OutcomeSU\n",
       "     diff_robust_(Intercept) diff_robust_Age diff_robust_IgE\n",
       "     diff_robust_OutcomeSU"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "“number of rows of result is not a multiple of vector length (arg 2)”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 25</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>taxon</th><th scope=col>lfc_(Intercept)</th><th scope=col>lfc_Age</th><th scope=col>lfc_IgE</th><th scope=col>lfc_OutcomeSU</th><th scope=col>p_(Intercept)</th><th scope=col>p_Age</th><th scope=col>p_IgE</th><th scope=col>p_OutcomeSU</th><th scope=col>q_(Intercept)</th><th scope=col>⋯</th><th scope=col>diff_IgE</th><th scope=col>diff_OutcomeSU</th><th scope=col>passed_ss_(Intercept)</th><th scope=col>passed_ss_Age</th><th scope=col>passed_ss_IgE</th><th scope=col>passed_ss_OutcomeSU</th><th scope=col>diff_robust_(Intercept)</th><th scope=col>diff_robust_Age</th><th scope=col>diff_robust_IgE</th><th scope=col>diff_robust_OutcomeSU</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 25\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " taxon & lfc\\_(Intercept) & lfc\\_Age & lfc\\_IgE & lfc\\_OutcomeSU & p\\_(Intercept) & p\\_Age & p\\_IgE & p\\_OutcomeSU & q\\_(Intercept) & ⋯ & diff\\_IgE & diff\\_OutcomeSU & passed\\_ss\\_(Intercept) & passed\\_ss\\_Age & passed\\_ss\\_IgE & passed\\_ss\\_OutcomeSU & diff\\_robust\\_(Intercept) & diff\\_robust\\_Age & diff\\_robust\\_IgE & diff\\_robust\\_OutcomeSU\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 25\n",
       "\n",
       "| taxon &lt;chr&gt; | lfc_(Intercept) &lt;dbl&gt; | lfc_Age &lt;dbl&gt; | lfc_IgE &lt;dbl&gt; | lfc_OutcomeSU &lt;dbl&gt; | p_(Intercept) &lt;dbl&gt; | p_Age &lt;dbl&gt; | p_IgE &lt;dbl&gt; | p_OutcomeSU &lt;dbl&gt; | q_(Intercept) &lt;dbl&gt; | ⋯ ⋯ | diff_IgE &lt;lgl&gt; | diff_OutcomeSU &lt;lgl&gt; | passed_ss_(Intercept) &lt;lgl&gt; | passed_ss_Age &lt;lgl&gt; | passed_ss_IgE &lt;lgl&gt; | passed_ss_OutcomeSU &lt;lgl&gt; | diff_robust_(Intercept) &lt;lgl&gt; | diff_robust_Age &lt;lgl&gt; | diff_robust_IgE &lt;lgl&gt; | diff_robust_OutcomeSU &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     taxon lfc_(Intercept) lfc_Age lfc_IgE lfc_OutcomeSU p_(Intercept) p_Age\n",
       "     p_IgE p_OutcomeSU q_(Intercept) ⋯ diff_IgE diff_OutcomeSU\n",
       "     passed_ss_(Intercept) passed_ss_Age passed_ss_IgE passed_ss_OutcomeSU\n",
       "     diff_robust_(Intercept) diff_robust_Age diff_robust_IgE\n",
       "     diff_robust_OutcomeSU"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# passed sensitivity filter and is significant \n",
    "stool_2cov_model %>%\n",
    "    filter(diff_robust_OutcomeSU == TRUE)\n",
    "\n",
    "# passed sensitivity analysis with a relaxed filter\n",
    "stool_2cov_model %>%\n",
    "    filter(q_OutcomeSU < 0.1 & passed_ss_OutcomeSU == TRUE)\n",
    "\n",
    "# significant with a relaxed filter\n",
    "stool_2cov_model %>%\n",
    "    filter(q_OutcomeSU < 0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97f926aa-8d5d-476a-a455-235f15853491",
   "metadata": {},
   "source": [
    "## Visualizing Differences in Abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c1e4a74c-d811-44b0-8362-a28e86f128fa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table style=\"width: 100%;\"><tr><td>ancombc2 {ANCOMBC}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Analysis of Compositions of Microbiomes with Bias Correction 2\n",
       "(ANCOM-BC2)</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>Determine taxa whose absolute abundances, per unit volume, of\n",
       "the ecosystem (e.g., gut) are significantly different with changes in the\n",
       "covariate of interest (e.g., group). The current version of\n",
       "<code>ancombc2</code> function implements Analysis of Compositions of Microbiomes\n",
       "with Bias Correction (ANCOM-BC2) in cross-sectional and repeated measurements\n",
       "data. In addition to the two-group comparison, ANCOM-BC2 also supports\n",
       "testing for continuous covariates and multi-group comparisons,\n",
       "including the global test, pairwise directional test, Dunnett's type of\n",
       "test, and trend test.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre><code class='language-R'>ancombc2(\n",
       "  data,\n",
       "  taxa_are_rows = TRUE,\n",
       "  assay.type = assay_name,\n",
       "  assay_name = \"counts\",\n",
       "  rank = tax_level,\n",
       "  tax_level = NULL,\n",
       "  aggregate_data = NULL,\n",
       "  meta_data = NULL,\n",
       "  fix_formula,\n",
       "  rand_formula = NULL,\n",
       "  p_adj_method = \"holm\",\n",
       "  pseudo = 0,\n",
       "  pseudo_sens = TRUE,\n",
       "  prv_cut = 0.1,\n",
       "  lib_cut = 0,\n",
       "  s0_perc = 0.05,\n",
       "  group = NULL,\n",
       "  struc_zero = FALSE,\n",
       "  neg_lb = FALSE,\n",
       "  alpha = 0.05,\n",
       "  n_cl = 1,\n",
       "  verbose = TRUE,\n",
       "  global = FALSE,\n",
       "  pairwise = FALSE,\n",
       "  dunnet = FALSE,\n",
       "  trend = FALSE,\n",
       "  iter_control = list(tol = 0.01, max_iter = 20, verbose = FALSE),\n",
       "  em_control = list(tol = 1e-05, max_iter = 100),\n",
       "  lme_control = lme4::lmerControl(),\n",
       "  mdfdr_control = list(fwer_ctrl_method = \"holm\", B = 100),\n",
       "  trend_control = list(contrast = NULL, node = NULL, solver = \"ECOS\", B = 100)\n",
       ")\n",
       "</code></pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table role = \"presentation\">\n",
       "<tr><td><code id=\"data\">data</code></td>\n",
       "<td>\n",
       "<p>the input data. The <code>data</code> parameter should be either a\n",
       "<code>matrix</code>, <code>data.frame</code>, <code>phyloseq</code> or a <code>TreeSummarizedExperiment</code>\n",
       "object. Both <code>phyloseq</code> and <code>TreeSummarizedExperiment</code> objects\n",
       "consist of a feature table (microbial count table), a sample metadata table,\n",
       "a taxonomy table (optional), and a phylogenetic tree (optional).\n",
       "If a <code>matrix</code> or <code>data.frame</code> is provided, ensure that the row\n",
       "names of the <code>metadata</code> match the sample names (column names if\n",
       "<code>taxa_are_rows</code> is TRUE, and row names otherwise) in <code>data</code>.\n",
       "if a <code>phyloseq</code> or a <code>TreeSummarizedExperiment</code> is used, this\n",
       "standard has already been enforced. For detailed information, refer to\n",
       "<code>?phyloseq::phyloseq</code> or\n",
       "<code>?TreeSummarizedExperiment::TreeSummarizedExperiment</code>.\n",
       "It is recommended to use low taxonomic levels, such as OTU or species level,\n",
       "as the estimation of sampling fractions requires a large number of taxa.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"taxa_are_rows\">taxa_are_rows</code></td>\n",
       "<td>\n",
       "<p>logical. Whether taxa are positioned in the rows of the\n",
       "feature table. Default is TRUE.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"assay.type\">assay.type</code></td>\n",
       "<td>\n",
       "<p>alias for <code>assay_name</code>.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"assay_name\">assay_name</code></td>\n",
       "<td>\n",
       "<p>character. Name of the count table in the data object\n",
       "(only applicable if data object is a <code>(Tree)SummarizedExperiment</code>).\n",
       "Default is &quot;counts&quot;.\n",
       "See <code>?SummarizedExperiment::assay</code> for more details.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"rank\">rank</code></td>\n",
       "<td>\n",
       "<p>alias for <code>tax_level</code>.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"tax_level\">tax_level</code></td>\n",
       "<td>\n",
       "<p>character. The taxonomic or non taxonomic(rowData) level of interest. The input data\n",
       "can be analyzed at any taxonomic or rowData level without prior agglomeration.\n",
       "Note that <code>tax_level</code> must be a value from <code>taxonomyRanks</code> or <code>rowData</code>, which\n",
       "includes &quot;Kingdom&quot;, &quot;Phylum&quot; &quot;Class&quot;, &quot;Order&quot;, &quot;Family&quot; &quot;Genus&quot; &quot;Species&quot; etc.\n",
       "See <code>?mia::taxonomyRanks</code> for more details.\n",
       "Default is NULL, i.e., do not perform agglomeration, and the\n",
       "ANCOM-BC2 analysis will be performed at the lowest taxonomic level of the\n",
       "input <code>data</code>.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"aggregate_data\">aggregate_data</code></td>\n",
       "<td>\n",
       "<p>The abundance data that has been aggregated to the desired\n",
       "taxonomic level. This parameter is required only when the input data is in\n",
       "<code>matrix</code> or <code>data.frame</code> format. For <code>phyloseq</code> or <code>TreeSummarizedExperiment</code>\n",
       "data, aggregation is performed by specifying the <code>tax_level</code> parameter.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"meta_data\">meta_data</code></td>\n",
       "<td>\n",
       "<p>a <code>data.frame</code> containing sample metadata.\n",
       "This parameter is mandatory when the input <code>data</code> is a generic\n",
       "<code>matrix</code> or <code>data.frame</code>. Ensure that the row names of the <code>metadata</code> match the\n",
       "sample names (column names if <code>taxa_are_rows</code> is TRUE, and row names\n",
       "otherwise) in <code>data</code>.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"fix_formula\">fix_formula</code></td>\n",
       "<td>\n",
       "<p>the character string expresses how the microbial absolute\n",
       "abundances for each taxon depend on the fixed effects in metadata. When\n",
       "specifying the <code>fix_formula</code>, make sure to include the <code>group</code>\n",
       "variable in the formula if it is not NULL.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"rand_formula\">rand_formula</code></td>\n",
       "<td>\n",
       "<p>the character string expresses how the microbial absolute\n",
       "abundances for each taxon depend on the random effects in metadata. ANCOM-BC2\n",
       "follows the <code>lmerTest</code> package in formulating the random effects. See\n",
       "<code>?lmerTest::lmer</code> for more details. Default is <code>NULL</code>.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"p_adj_method\">p_adj_method</code></td>\n",
       "<td>\n",
       "<p>character. method to adjust p-values. Default is &quot;holm&quot;.\n",
       "Options include &quot;holm&quot;, &quot;hochberg&quot;, &quot;hommel&quot;, &quot;bonferroni&quot;, &quot;BH&quot;, &quot;BY&quot;,\n",
       "&quot;fdr&quot;, &quot;none&quot;. See <code>?stats::p.adjust</code> for more details.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"pseudo\">pseudo</code></td>\n",
       "<td>\n",
       "<p>A small positive value (default: 1) added to all counts\n",
       "before log transformation to avoid numerical issues caused by log(0).</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"pseudo_sens\">pseudo_sens</code></td>\n",
       "<td>\n",
       "<p>logical. Whether to perform the sensitivity analysis to\n",
       "the pseudo-count addition. Default is <code>TRUE</code>. While ANCOM-BC2 utilizes\n",
       "complete data (nonzero counts) by default for its analysis, a comprehensive\n",
       "evaluation of result robustness is performed by assessing how pseudo-count\n",
       "addition to zeros may affect the outcomes. For a detailed discussion on this\n",
       "sensitivity analysis, refer to the <code>Details</code> section.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"prv_cut\">prv_cut</code></td>\n",
       "<td>\n",
       "<p>a numerical fraction between 0 and 1. Taxa with prevalences\n",
       "(the proportion of samples in which the taxon is present)\n",
       "less than <code>prv_cut</code> will be excluded in the analysis. For example,\n",
       "if there are 100 samples, and a taxon has nonzero counts present in less than\n",
       "100*prv_cut samples, it will not be considered in the analysis.\n",
       "Default is 0.10.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"lib_cut\">lib_cut</code></td>\n",
       "<td>\n",
       "<p>a numerical threshold for filtering samples based on library\n",
       "sizes. Samples with library sizes less than <code>lib_cut</code> will be\n",
       "excluded in the analysis. Default is 0, i.e. do not discard any sample.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"s0_perc\">s0_perc</code></td>\n",
       "<td>\n",
       "<p>a numerical fraction between 0 and 1. Inspired by\n",
       "<a href=\"https://doi.org/10.1073/pnas.091062498\">Significance\n",
       "Analysis of Microarrays (SAM)</a> methodology, a small positive constant is\n",
       "added to the denominator of ANCOM-BC2 test statistic corresponding to\n",
       "each taxon to avoid the significance due to extremely small standard errors,\n",
       "especially for rare taxa. This small positive constant is chosen as\n",
       "<code>s0_perc</code>-th percentile of standard error values for each fixed effect.\n",
       "Default is 0.05 (5th percentile).</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"group\">group</code></td>\n",
       "<td>\n",
       "<p>character. the name of the group variable in metadata.\n",
       "The <code>group</code> parameter should be a character string representing the name\n",
       "of the group variable in the metadata. The <code>group</code> variable should be\n",
       "discrete, meaning it consists of categorical values. Specifying the\n",
       "<code>group</code> variable is required if you are interested in detecting\n",
       "structural zeros and performing performing multi-group comparisons (global\n",
       "test, pairwise directional test, Dunnett's type of test, and trend test).\n",
       "However, if these analyses are not of interest to you, you can leave the\n",
       "<code>group</code> parameter as NULL. If the <code>group</code> variable of interest\n",
       "contains only two categories, you can also leave the <code>group</code> parameter\n",
       "as NULL. Default is NULL.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"struc_zero\">struc_zero</code></td>\n",
       "<td>\n",
       "<p>logical. Whether to detect structural zeros based on\n",
       "<code>group</code>. Default is FALSE. See <code>Details</code> for\n",
       "a more comprehensive discussion on structural zeros.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"neg_lb\">neg_lb</code></td>\n",
       "<td>\n",
       "<p>logical. Whether to classify a taxon as a structural zero using\n",
       "its asymptotic lower bound. Default is FALSE.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"alpha\">alpha</code></td>\n",
       "<td>\n",
       "<p>numeric. Level of significance. Default is 0.05.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"n_cl\">n_cl</code></td>\n",
       "<td>\n",
       "<p>numeric. The number of nodes to be forked. For details, see\n",
       "<code>?parallel::makeCluster</code>. Default is 1 (no parallel computing).</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"verbose\">verbose</code></td>\n",
       "<td>\n",
       "<p>logical. Whether to generate verbose output during the\n",
       "ANCOM-BC2 fitting process. Default is FALSE.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"global\">global</code></td>\n",
       "<td>\n",
       "<p>logical. Whether to perform the global test. Default is FALSE.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"pairwise\">pairwise</code></td>\n",
       "<td>\n",
       "<p>logical. Whether to perform the pairwise directional test.\n",
       "Default is FALSE.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"dunnet\">dunnet</code></td>\n",
       "<td>\n",
       "<p>logical. Whether to perform the Dunnett's type of test.\n",
       "Default is FALSE.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"trend\">trend</code></td>\n",
       "<td>\n",
       "<p>logical. Whether to perform trend test. Default is FALSE.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"iter_control\">iter_control</code></td>\n",
       "<td>\n",
       "<p>a named list of control parameters for the iterative\n",
       "MLE or RMEL algorithm, including 1) <code>tol</code>: the iteration convergence\n",
       "tolerance (default is 1e-02), 2) <code>max_iter</code>: the maximum number of\n",
       "iterations (default is 20), and 3)<code>verbose</code>: whether to show the verbose\n",
       "output (default is FALSE).</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"em_control\">em_control</code></td>\n",
       "<td>\n",
       "<p>a named list of control parameters for the E-M algorithm,\n",
       "including 1) <code>tol</code>: the iteration convergence tolerance\n",
       "(default is 1e-05) and 2) <code>max_iter</code>: the maximum number of iterations\n",
       "(default is 100).</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"lme_control\">lme_control</code></td>\n",
       "<td>\n",
       "<p>a list of control parameters for mixed model fitting.\n",
       "See <code>?lme4::lmerControl</code> for details.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"mdfdr_control\">mdfdr_control</code></td>\n",
       "<td>\n",
       "<p>a named list of control parameters for mixed directional\n",
       "false discover rate (mdFDR), including 1) <code>fwer_ctrl_method</code>: family\n",
       "wise error (FWER) controlling procedure, such as &quot;holm&quot;, &quot;hochberg&quot;,\n",
       "&quot;bonferroni&quot;, etc (default is &quot;holm&quot;) and 2) <code>B</code>: the number of\n",
       "bootstrap samples (default is 100). Increase <code>B</code> will lead to a more\n",
       "accurate p-values. See <code>Details</code> for a more comprehensive discussion on\n",
       "mdFDR.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"trend_control\">trend_control</code></td>\n",
       "<td>\n",
       "<p>a named list of control parameters for the trend test,\n",
       "including 1) <code>contrast</code>: the list of contrast matrices for\n",
       "constructing inequalities, 2) <code>node</code>: the list of positions for the\n",
       "nodal parameter, 3) <code>solver</code>: a string indicating the solver to use\n",
       "(default is &quot;ECOS&quot;), and 4) <code>B</code>: the number of bootstrap samples\n",
       "(default is 100). Increase <code>B</code> will lead to a more accurate p-values.\n",
       "See <code>vignette</code> for the corresponding trend test examples.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>A taxon is considered to have structural zeros in some (&gt;=1)\n",
       "groups if it is completely (or nearly completely) missing in these groups.\n",
       "For instance, suppose there are three groups: g1, g2, and g3.\n",
       "If the counts of taxon A in g1 are 0 but nonzero in g2 and g3,\n",
       "then taxon A will be considered to contain structural zeros in g1.\n",
       "In this example, taxon A is declared to be differentially abundant between\n",
       "g1 and g2, g1 and g3, and consequently, it is globally differentially\n",
       "abundant with respect to this group variable.\n",
       "Such taxa are not further analyzed using ANCOM-BC2, but the results are\n",
       "summarized in the overall summary. For more details about the structural\n",
       "zeros, please go to the\n",
       "<a href=\"https://doi.org/10.3389/fmicb.2017.02114\">ANCOM-II</a> paper.\n",
       "Setting <code>neg_lb = TRUE</code> indicates that you are using both criteria\n",
       "stated in section 3.2 of\n",
       "<a href=\"https://doi.org/10.3389/fmicb.2017.02114\">ANCOM-II</a>\n",
       "to detect structural zeros; otherwise, the algorithm will only use the\n",
       "equation 1 in section 3.2 for declaring structural zeros. Generally, it is\n",
       "recommended to set <code>neg_lb = TRUE</code> when the sample size per group is\n",
       "relatively large (e.g. &gt; 30).\n",
       "</p>\n",
       "<p>Like other differential abundance analysis methods, ANCOM-BC2 applies a log\n",
       "transformation to the observed counts. However, the presence of zero counts\n",
       "poses a challenge, and researchers often consider adding a pseudo-count before\n",
       "the log transformation. However, it has been shown that the choice of\n",
       "pseudo-count can impact the results and lead to an inflated false positive\n",
       "rate (<a href=\"https://doi.org/10.1038/nmeth.2897\">Costea et al. (2014)</a>;\n",
       "<a href=\"https://doi.org/10.1038/nmeth.2898\">Paulson, Bravo, and Pop (2014)</a>).\n",
       "To address this issue, we conduct a sensitivity analysis to assess the impact\n",
       "of different pseudo-counts on zero counts for each taxon. This involves\n",
       "adding various pseudo-count values (0.1, 0.5, and 1) to the feature table and\n",
       "re-running the ANCOM-BC2 algorithm. The sensitivity score for each taxon is\n",
       "calculated as the proportion of times that the q-value exceeds the specified\n",
       "significance level (alpha). If all q-values consistently show significance or\n",
       "nonsignificance across different pseudo-counts and are consistent with the\n",
       "results obtained without adding pseudo-counts to zero counts (using the\n",
       "default settings), then the taxon is considered not sensitive to the\n",
       "pseudo-count addition.\n",
       "</p>\n",
       "<p>When performning pairwise directional (or Dunnett's type of) test, the mixed\n",
       "directional false discover rate (mdFDR) should be taken into account.\n",
       "The mdFDR is the combination of false discovery rate due to multiple testing,\n",
       "multiple pairwise comparisons, and directional tests within each pairwise\n",
       "comparison. For example, suppose we have five taxa and three experimental\n",
       "groups: g1, g2, and g3. Thus, we are performing five tests corresponding to\n",
       "five taxa. For each taxon, we are also conducting three pairwise comparisons\n",
       "(g1 vs. g2, g2 vs. g3, and g1 vs. g3). Within each pairwise comparison,\n",
       "we wish to determine if the abundance has increased or decreased or did not\n",
       "change (direction of the effect size). Errors could occur in each step.\n",
       "The overall false discovery rate is controlled by the mdFDR methodology we\n",
       "adopted from\n",
       "<a href=\"https://doi.org/10.1111/j.1541-0420.2009.01292.x\">Guo, Sarkar, and Peddada (2010)</a> and\n",
       "<a href=\"https://doi.org/10.1186/s12859-016-0937-5\">Grandhi, Guo, and Peddada (2016)</a>.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>a <code>list</code> with components:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li> <p><code>feature_table</code>, a <code>data.frame</code> of pre-processed\n",
       "(based on <code>prv_cut</code> and <code>lib_cut</code>) microbial count table.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>bias_correct_log_table</code>, a <code>data.frame</code> of\n",
       "bias-corrected log abundance table.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>ss_tab</code>, a <code>data.frame</code> of sensitivity scores for\n",
       "pseudo-count addition to 0s.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>zero_ind</code>, a logical <code>data.frame</code> with TRUE\n",
       "indicating the taxon is detected to contain structural zeros in\n",
       "some specific groups.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>samp_frac</code>, a numeric vector of estimated sampling\n",
       "fractions in log scale (natural log).\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>delta_em</code>, estimated sample-specific biases\n",
       "through E-M algorithm.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>delta_wls</code>, estimated sample-specific biases through\n",
       "weighted least squares (WLS) algorithm.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>res</code>,  a <code>data.frame</code> containing ANCOM-BC2 primary\n",
       "result:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li><p> columns started with <code>lfc</code>: log fold changes\n",
       "obtained from the ANCOM-BC2 log-linear (natural log) model.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>se</code>: standard errors (SEs) of\n",
       "<code>lfc</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>W</code>: test statistics.\n",
       "<code>W = lfc/se</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>p</code>: p-values. P-values are\n",
       "obtained from two-sided Z-test using the test statistic <code>W</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>q</code>: adjusted p-values.\n",
       "Adjusted p-values are obtained by applying <code>p_adj_method</code>\n",
       "to <code>p</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>diff</code>: TRUE if the\n",
       "taxon is significant (has <code>q</code> less than <code>alpha</code>).\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>passed_ss</code>: TRUE if the\n",
       "taxon passed the sensitivity analysis, i.e., adding different\n",
       "pseudo-counts to 0s would not change the results.\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "</li>\n",
       "<li><p> columns started with <code>diff_robust</code>: TRUE if the\n",
       "taxon is significant (has <code>q</code> less than <code>alpha</code>) and robust\n",
       "in the sensitivity analysis (<code>passed_ss</code> is <code>TRUE</code>).\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>res_global</code>,  a <code>data.frame</code> containing ANCOM-BC2\n",
       "global test result for the variable specified in <code>group</code>,\n",
       "each column is:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li> <p><code>W</code>, test statistics.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>p_val</code>, p-values, which are obtained from two-sided\n",
       "Chi-square test using <code>W</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>q_val</code>, adjusted p-values. Adjusted p-values are\n",
       "obtained by applying <code>p_adj_method</code> to <code>p_val</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>diff_abn</code>, A logical vector. TRUE if the taxon has\n",
       "<code>q_val</code> less than <code>alpha</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>passed_ss</code>, A logical vector. TRUE if the taxon has\n",
       "passed the sensitivity analysis.\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "</li>\n",
       "<li> <p><code>res_pair</code>,  a <code>data.frame</code> containing ANCOM-BC2\n",
       "pairwise directional test result for the variable specified in\n",
       "<code>group</code>:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li><p> columns started with <code>lfc</code>: log fold changes.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>se</code>: standard errors (SEs).\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>W</code>: test statistics.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>p</code>: p-values.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>q</code>: adjusted p-values.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>diff</code>: TRUE if the\n",
       "taxon is significant (has <code>q</code> less than <code>alpha</code>).\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>passed_ss</code>: TRUE if the\n",
       "taxon has passed the sensitivity analysis.\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "</li>\n",
       "<li> <p><code>res_dunn</code>,  a <code>data.frame</code> containing ANCOM-BC2\n",
       "Dunnett's type of test result for the variable specified in\n",
       "<code>group</code>:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li><p> columns started with <code>lfc</code>: log fold changes.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>se</code>: standard errors (SEs).\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>W</code>: test statistics.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>p</code>: p-values.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>q</code>: adjusted p-values.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>diff</code>: TRUE if the\n",
       "taxon is significant (has <code>q</code> less than <code>alpha</code>).\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>passed_ss</code>: TRUE if the\n",
       "taxon has passed the sensitivity analysis.\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "</li>\n",
       "<li> <p><code>res_trend</code>,  a <code>data.frame</code> containing ANCOM-BC2\n",
       "trend test result for the variable specified in\n",
       "<code>group</code>:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li><p> columns started with <code>lfc</code>: log fold changes.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> columns started with <code>se</code>: standard errors (SEs).\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>W</code>: test statistics.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>p_val</code>: p-values.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>q_val</code>: adjusted p-values.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>diff_abn</code>: TRUE if the\n",
       "taxon is significant (has <code>q</code> less than <code>alpha</code>).\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>passed_ss</code>, A logical vector. TRUE if the taxon has\n",
       "passed the sensitivity analysis.\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "</li></ul>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>Huang Lin\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>Kaul A, Mandal S, Davidov O, Peddada SD (2017).\n",
       "&ldquo;Analysis of microbiome data in the presence of excess zeros.&rdquo;\n",
       "<em>Frontiers in microbiology</em>, <b>8</b>, 2114.\n",
       "</p>\n",
       "<p>Lin H, Peddada SD (2020).\n",
       "&ldquo;Analysis of compositions of microbiomes with bias correction.&rdquo;\n",
       "<em>Nature communications</em>, <b>11</b>(1), 1&ndash;11.\n",
       "</p>\n",
       "<p>Tusher VG, Tibshirani R, Chu G (2001).\n",
       "&ldquo;Significance analysis of microarrays applied to the ionizing radiation response.&rdquo;\n",
       "<em>Proceedings of the National Academy of Sciences</em>, <b>98</b>(9), 5116&ndash;5121.\n",
       "</p>\n",
       "<p>Costea PI, Zeller G, Sunagawa S, Bork P (2014).\n",
       "&ldquo;A fair comparison.&rdquo;\n",
       "<em>Nature methods</em>, <b>11</b>(4), 359&ndash;359.\n",
       "</p>\n",
       "<p>Paulson JN, Bravo HC, Pop M (2014).\n",
       "&ldquo;Reply to:&quot; a fair comparison&quot;.&rdquo;\n",
       "<em>Nature methods</em>, <b>11</b>(4), 359&ndash;360.\n",
       "</p>\n",
       "<p>Guo W, Sarkar SK, Peddada SD (2010).\n",
       "&ldquo;Controlling false discoveries in multidimensional directional decisions, with applications to gene expression data on ordered categories.&rdquo;\n",
       "<em>Biometrics</em>, <b>66</b>(2), 485&ndash;492.\n",
       "</p>\n",
       "<p>Grandhi A, Guo W, Peddada SD (2016).\n",
       "&ldquo;A multiple testing procedure for multi-dimensional pairwise comparisons with application to gene expression studies.&rdquo;\n",
       "<em>BMC bioinformatics</em>, <b>17</b>(1), 1&ndash;12.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "<p><code>ancom</code> <code>ancombc</code>\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre><code class='language-R'>library(ANCOMBC)\n",
       "if (requireNamespace(\"microbiome\", quietly = TRUE)) {\n",
       "    data(dietswap, package = \"microbiome\")\n",
       "\n",
       "    # run ancombc2 function\n",
       "    set.seed(123)\n",
       "    # Note that setting max_iter = 1 and B = 1 is only for the sake of speed\n",
       "    # Use default or larger values for max_iter and B for better performance\n",
       "    out = ancombc2(data = dietswap, tax_level = \"Family\",\n",
       "                   fix_formula = \"nationality + timepoint + bmi_group\",\n",
       "                   rand_formula = NULL,\n",
       "                   p_adj_method = \"holm\", pseudo_sens = TRUE,\n",
       "                   prv_cut = 0.10, lib_cut = 1000, s0_perc = 0.05,\n",
       "                   group = \"bmi_group\", struc_zero = TRUE, neg_lb = TRUE,\n",
       "                   alpha = 0.05, n_cl = 1, verbose = TRUE,\n",
       "                   global = TRUE, pairwise = TRUE, dunnet = TRUE, trend = TRUE,\n",
       "                   iter_control = list(tol = 1e-2, max_iter = 1, verbose = TRUE),\n",
       "                   em_control = list(tol = 1e-5, max_iter = 1),\n",
       "                   lme_control = lme4::lmerControl(),\n",
       "                   mdfdr_control = list(fwer_ctrl_method = \"holm\", B = 1),\n",
       "                   trend_control = list(contrast =\n",
       "                                            list(matrix(c(1, 0, -1, 1),\n",
       "                                                        nrow = 2,\n",
       "                                                        byrow = TRUE)),\n",
       "                                        node = list(2),\n",
       "                                        solver = \"ECOS\",\n",
       "                                        B = 1))\n",
       "    res_prim = out$res\n",
       "    res_global = out$res_global\n",
       "    res_pair = out$res_pair\n",
       "    res_dunn = out$res_dunn\n",
       "    res_trend = out$res_trend\n",
       "} else {\n",
       "    message(\"The 'microbiome' package is not installed. Please install it to use this example.\")\n",
       "}\n",
       "\n",
       "</code></pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>ANCOMBC</em> version 2.10.1 ]</div></main>\n",
       "\n",
       "</div>"
      ],
      "text/latex": [
       "\\HeaderA{ancombc2}{Analysis of Compositions of Microbiomes with Bias Correction 2 (ANCOM-BC2)}{ancombc2}\n",
       "%\n",
       "\\begin{Description}\n",
       "Determine taxa whose absolute abundances, per unit volume, of\n",
       "the ecosystem (e.g., gut) are significantly different with changes in the\n",
       "covariate of interest (e.g., group). The current version of\n",
       "\\code{ancombc2} function implements Analysis of Compositions of Microbiomes\n",
       "with Bias Correction (ANCOM-BC2) in cross-sectional and repeated measurements\n",
       "data. In addition to the two-group comparison, ANCOM-BC2 also supports\n",
       "testing for continuous covariates and multi-group comparisons,\n",
       "including the global test, pairwise directional test, Dunnett's type of\n",
       "test, and trend test.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "ancombc2(\n",
       "  data,\n",
       "  taxa_are_rows = TRUE,\n",
       "  assay.type = assay_name,\n",
       "  assay_name = \"counts\",\n",
       "  rank = tax_level,\n",
       "  tax_level = NULL,\n",
       "  aggregate_data = NULL,\n",
       "  meta_data = NULL,\n",
       "  fix_formula,\n",
       "  rand_formula = NULL,\n",
       "  p_adj_method = \"holm\",\n",
       "  pseudo = 0,\n",
       "  pseudo_sens = TRUE,\n",
       "  prv_cut = 0.1,\n",
       "  lib_cut = 0,\n",
       "  s0_perc = 0.05,\n",
       "  group = NULL,\n",
       "  struc_zero = FALSE,\n",
       "  neg_lb = FALSE,\n",
       "  alpha = 0.05,\n",
       "  n_cl = 1,\n",
       "  verbose = TRUE,\n",
       "  global = FALSE,\n",
       "  pairwise = FALSE,\n",
       "  dunnet = FALSE,\n",
       "  trend = FALSE,\n",
       "  iter_control = list(tol = 0.01, max_iter = 20, verbose = FALSE),\n",
       "  em_control = list(tol = 1e-05, max_iter = 100),\n",
       "  lme_control = lme4::lmerControl(),\n",
       "  mdfdr_control = list(fwer_ctrl_method = \"holm\", B = 100),\n",
       "  trend_control = list(contrast = NULL, node = NULL, solver = \"ECOS\", B = 100)\n",
       ")\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{data}] the input data. The \\code{data} parameter should be either a\n",
       "\\code{matrix}, \\code{data.frame}, \\code{phyloseq} or a \\code{TreeSummarizedExperiment}\n",
       "object. Both \\code{phyloseq} and \\code{TreeSummarizedExperiment} objects\n",
       "consist of a feature table (microbial count table), a sample metadata table,\n",
       "a taxonomy table (optional), and a phylogenetic tree (optional).\n",
       "If a \\code{matrix} or \\code{data.frame} is provided, ensure that the row\n",
       "names of the \\code{metadata} match the sample names (column names if\n",
       "\\code{taxa\\_are\\_rows} is TRUE, and row names otherwise) in \\code{data}.\n",
       "if a \\code{phyloseq} or a \\code{TreeSummarizedExperiment} is used, this\n",
       "standard has already been enforced. For detailed information, refer to\n",
       "\\code{?phyloseq::phyloseq} or\n",
       "\\code{?TreeSummarizedExperiment::TreeSummarizedExperiment}.\n",
       "It is recommended to use low taxonomic levels, such as OTU or species level,\n",
       "as the estimation of sampling fractions requires a large number of taxa.\n",
       "\n",
       "\\item[\\code{taxa\\_are\\_rows}] logical. Whether taxa are positioned in the rows of the\n",
       "feature table. Default is TRUE.\n",
       "\n",
       "\\item[\\code{assay.type}] alias for \\code{assay\\_name}.\n",
       "\n",
       "\\item[\\code{assay\\_name}] character. Name of the count table in the data object\n",
       "(only applicable if data object is a \\code{(Tree)SummarizedExperiment}).\n",
       "Default is \"counts\".\n",
       "See \\code{?SummarizedExperiment::assay} for more details.\n",
       "\n",
       "\\item[\\code{rank}] alias for \\code{tax\\_level}.\n",
       "\n",
       "\\item[\\code{tax\\_level}] character. The taxonomic or non taxonomic(rowData) level of interest. The input data\n",
       "can be analyzed at any taxonomic or rowData level without prior agglomeration.\n",
       "Note that \\code{tax\\_level} must be a value from \\code{taxonomyRanks} or \\code{rowData}, which\n",
       "includes \"Kingdom\", \"Phylum\" \"Class\", \"Order\", \"Family\" \"Genus\" \"Species\" etc.\n",
       "See \\code{?mia::taxonomyRanks} for more details.\n",
       "Default is NULL, i.e., do not perform agglomeration, and the\n",
       "ANCOM-BC2 analysis will be performed at the lowest taxonomic level of the\n",
       "input \\code{data}.\n",
       "\n",
       "\\item[\\code{aggregate\\_data}] The abundance data that has been aggregated to the desired\n",
       "taxonomic level. This parameter is required only when the input data is in\n",
       "\\code{matrix} or \\code{data.frame} format. For \\code{phyloseq} or \\code{TreeSummarizedExperiment}\n",
       "data, aggregation is performed by specifying the \\code{tax\\_level} parameter.\n",
       "\n",
       "\\item[\\code{meta\\_data}] a \\code{data.frame} containing sample metadata.\n",
       "This parameter is mandatory when the input \\code{data} is a generic\n",
       "\\code{matrix} or \\code{data.frame}. Ensure that the row names of the \\code{metadata} match the\n",
       "sample names (column names if \\code{taxa\\_are\\_rows} is TRUE, and row names\n",
       "otherwise) in \\code{data}.\n",
       "\n",
       "\\item[\\code{fix\\_formula}] the character string expresses how the microbial absolute\n",
       "abundances for each taxon depend on the fixed effects in metadata. When\n",
       "specifying the \\code{fix\\_formula}, make sure to include the \\code{group}\n",
       "variable in the formula if it is not NULL.\n",
       "\n",
       "\\item[\\code{rand\\_formula}] the character string expresses how the microbial absolute\n",
       "abundances for each taxon depend on the random effects in metadata. ANCOM-BC2\n",
       "follows the \\code{lmerTest} package in formulating the random effects. See\n",
       "\\code{?lmerTest::lmer} for more details. Default is \\code{NULL}.\n",
       "\n",
       "\\item[\\code{p\\_adj\\_method}] character. method to adjust p-values. Default is \"holm\".\n",
       "Options include \"holm\", \"hochberg\", \"hommel\", \"bonferroni\", \"BH\", \"BY\",\n",
       "\"fdr\", \"none\". See \\code{?stats::p.adjust} for more details.\n",
       "\n",
       "\\item[\\code{pseudo}] A small positive value (default: 1) added to all counts\n",
       "before log transformation to avoid numerical issues caused by log(0).\n",
       "\n",
       "\\item[\\code{pseudo\\_sens}] logical. Whether to perform the sensitivity analysis to\n",
       "the pseudo-count addition. Default is \\code{TRUE}. While ANCOM-BC2 utilizes\n",
       "complete data (nonzero counts) by default for its analysis, a comprehensive\n",
       "evaluation of result robustness is performed by assessing how pseudo-count\n",
       "addition to zeros may affect the outcomes. For a detailed discussion on this\n",
       "sensitivity analysis, refer to the \\code{Details} section.\n",
       "\n",
       "\\item[\\code{prv\\_cut}] a numerical fraction between 0 and 1. Taxa with prevalences\n",
       "(the proportion of samples in which the taxon is present)\n",
       "less than \\code{prv\\_cut} will be excluded in the analysis. For example,\n",
       "if there are 100 samples, and a taxon has nonzero counts present in less than\n",
       "100*prv\\_cut samples, it will not be considered in the analysis.\n",
       "Default is 0.10.\n",
       "\n",
       "\\item[\\code{lib\\_cut}] a numerical threshold for filtering samples based on library\n",
       "sizes. Samples with library sizes less than \\code{lib\\_cut} will be\n",
       "excluded in the analysis. Default is 0, i.e. do not discard any sample.\n",
       "\n",
       "\\item[\\code{s0\\_perc}] a numerical fraction between 0 and 1. Inspired by\n",
       "\\Rhref{https://doi.org/10.1073/pnas.091062498}{Significance\n",
       "Analysis of Microarrays (SAM)} methodology, a small positive constant is\n",
       "added to the denominator of ANCOM-BC2 test statistic corresponding to\n",
       "each taxon to avoid the significance due to extremely small standard errors,\n",
       "especially for rare taxa. This small positive constant is chosen as\n",
       "\\code{s0\\_perc}-th percentile of standard error values for each fixed effect.\n",
       "Default is 0.05 (5th percentile).\n",
       "\n",
       "\\item[\\code{group}] character. the name of the group variable in metadata.\n",
       "The \\code{group} parameter should be a character string representing the name\n",
       "of the group variable in the metadata. The \\code{group} variable should be\n",
       "discrete, meaning it consists of categorical values. Specifying the\n",
       "\\code{group} variable is required if you are interested in detecting\n",
       "structural zeros and performing performing multi-group comparisons (global\n",
       "test, pairwise directional test, Dunnett's type of test, and trend test).\n",
       "However, if these analyses are not of interest to you, you can leave the\n",
       "\\code{group} parameter as NULL. If the \\code{group} variable of interest\n",
       "contains only two categories, you can also leave the \\code{group} parameter\n",
       "as NULL. Default is NULL.\n",
       "\n",
       "\\item[\\code{struc\\_zero}] logical. Whether to detect structural zeros based on\n",
       "\\code{group}. Default is FALSE. See \\code{Details} for\n",
       "a more comprehensive discussion on structural zeros.\n",
       "\n",
       "\\item[\\code{neg\\_lb}] logical. Whether to classify a taxon as a structural zero using\n",
       "its asymptotic lower bound. Default is FALSE.\n",
       "\n",
       "\\item[\\code{alpha}] numeric. Level of significance. Default is 0.05.\n",
       "\n",
       "\\item[\\code{n\\_cl}] numeric. The number of nodes to be forked. For details, see\n",
       "\\code{?parallel::makeCluster}. Default is 1 (no parallel computing).\n",
       "\n",
       "\\item[\\code{verbose}] logical. Whether to generate verbose output during the\n",
       "ANCOM-BC2 fitting process. Default is FALSE.\n",
       "\n",
       "\\item[\\code{global}] logical. Whether to perform the global test. Default is FALSE.\n",
       "\n",
       "\\item[\\code{pairwise}] logical. Whether to perform the pairwise directional test.\n",
       "Default is FALSE.\n",
       "\n",
       "\\item[\\code{dunnet}] logical. Whether to perform the Dunnett's type of test.\n",
       "Default is FALSE.\n",
       "\n",
       "\\item[\\code{trend}] logical. Whether to perform trend test. Default is FALSE.\n",
       "\n",
       "\\item[\\code{iter\\_control}] a named list of control parameters for the iterative\n",
       "MLE or RMEL algorithm, including 1) \\code{tol}: the iteration convergence\n",
       "tolerance (default is 1e-02), 2) \\code{max\\_iter}: the maximum number of\n",
       "iterations (default is 20), and 3)\\code{verbose}: whether to show the verbose\n",
       "output (default is FALSE).\n",
       "\n",
       "\\item[\\code{em\\_control}] a named list of control parameters for the E-M algorithm,\n",
       "including 1) \\code{tol}: the iteration convergence tolerance\n",
       "(default is 1e-05) and 2) \\code{max\\_iter}: the maximum number of iterations\n",
       "(default is 100).\n",
       "\n",
       "\\item[\\code{lme\\_control}] a list of control parameters for mixed model fitting.\n",
       "See \\code{?lme4::lmerControl} for details.\n",
       "\n",
       "\\item[\\code{mdfdr\\_control}] a named list of control parameters for mixed directional\n",
       "false discover rate (mdFDR), including 1) \\code{fwer\\_ctrl\\_method}: family\n",
       "wise error (FWER) controlling procedure, such as \"holm\", \"hochberg\",\n",
       "\"bonferroni\", etc (default is \"holm\") and 2) \\code{B}: the number of\n",
       "bootstrap samples (default is 100). Increase \\code{B} will lead to a more\n",
       "accurate p-values. See \\code{Details} for a more comprehensive discussion on\n",
       "mdFDR.\n",
       "\n",
       "\\item[\\code{trend\\_control}] a named list of control parameters for the trend test,\n",
       "including 1) \\code{contrast}: the list of contrast matrices for\n",
       "constructing inequalities, 2) \\code{node}: the list of positions for the\n",
       "nodal parameter, 3) \\code{solver}: a string indicating the solver to use\n",
       "(default is \"ECOS\"), and 4) \\code{B}: the number of bootstrap samples\n",
       "(default is 100). Increase \\code{B} will lead to a more accurate p-values.\n",
       "See \\code{vignette} for the corresponding trend test examples.\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\n",
       "A taxon is considered to have structural zeros in some (>=1)\n",
       "groups if it is completely (or nearly completely) missing in these groups.\n",
       "For instance, suppose there are three groups: g1, g2, and g3.\n",
       "If the counts of taxon A in g1 are 0 but nonzero in g2 and g3,\n",
       "then taxon A will be considered to contain structural zeros in g1.\n",
       "In this example, taxon A is declared to be differentially abundant between\n",
       "g1 and g2, g1 and g3, and consequently, it is globally differentially\n",
       "abundant with respect to this group variable.\n",
       "Such taxa are not further analyzed using ANCOM-BC2, but the results are\n",
       "summarized in the overall summary. For more details about the structural\n",
       "zeros, please go to the\n",
       "\\Rhref{https://doi.org/10.3389/fmicb.2017.02114}{ANCOM-II} paper.\n",
       "Setting \\code{neg\\_lb = TRUE} indicates that you are using both criteria\n",
       "stated in section 3.2 of\n",
       "\\Rhref{https://doi.org/10.3389/fmicb.2017.02114}{ANCOM-II}\n",
       "to detect structural zeros; otherwise, the algorithm will only use the\n",
       "equation 1 in section 3.2 for declaring structural zeros. Generally, it is\n",
       "recommended to set \\code{neg\\_lb = TRUE} when the sample size per group is\n",
       "relatively large (e.g. > 30).\n",
       "\n",
       "Like other differential abundance analysis methods, ANCOM-BC2 applies a log\n",
       "transformation to the observed counts. However, the presence of zero counts\n",
       "poses a challenge, and researchers often consider adding a pseudo-count before\n",
       "the log transformation. However, it has been shown that the choice of\n",
       "pseudo-count can impact the results and lead to an inflated false positive\n",
       "rate (\\Rhref{https://doi.org/10.1038/nmeth.2897}{Costea et al. (2014)};\n",
       "\\Rhref{https://doi.org/10.1038/nmeth.2898}{Paulson, Bravo, and Pop (2014)}).\n",
       "To address this issue, we conduct a sensitivity analysis to assess the impact\n",
       "of different pseudo-counts on zero counts for each taxon. This involves\n",
       "adding various pseudo-count values (0.1, 0.5, and 1) to the feature table and\n",
       "re-running the ANCOM-BC2 algorithm. The sensitivity score for each taxon is\n",
       "calculated as the proportion of times that the q-value exceeds the specified\n",
       "significance level (alpha). If all q-values consistently show significance or\n",
       "nonsignificance across different pseudo-counts and are consistent with the\n",
       "results obtained without adding pseudo-counts to zero counts (using the\n",
       "default settings), then the taxon is considered not sensitive to the\n",
       "pseudo-count addition.\n",
       "\n",
       "When performning pairwise directional (or Dunnett's type of) test, the mixed\n",
       "directional false discover rate (mdFDR) should be taken into account.\n",
       "The mdFDR is the combination of false discovery rate due to multiple testing,\n",
       "multiple pairwise comparisons, and directional tests within each pairwise\n",
       "comparison. For example, suppose we have five taxa and three experimental\n",
       "groups: g1, g2, and g3. Thus, we are performing five tests corresponding to\n",
       "five taxa. For each taxon, we are also conducting three pairwise comparisons\n",
       "(g1 vs. g2, g2 vs. g3, and g1 vs. g3). Within each pairwise comparison,\n",
       "we wish to determine if the abundance has increased or decreased or did not\n",
       "change (direction of the effect size). Errors could occur in each step.\n",
       "The overall false discovery rate is controlled by the mdFDR methodology we\n",
       "adopted from\n",
       "\\Rhref{https://doi.org/10.1111/j.1541-0420.2009.01292.x}{Guo, Sarkar, and Peddada (2010)} and\n",
       "\\Rhref{https://doi.org/10.1186/s12859-016-0937-5}{Grandhi, Guo, and Peddada (2016)}.\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "a \\code{list} with components:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{}  \\code{feature\\_table}, a \\code{data.frame} of pre-processed\n",
       "(based on \\code{prv\\_cut} and \\code{lib\\_cut}) microbial count table.\n",
       "\\item{}  \\code{bias\\_correct\\_log\\_table}, a \\code{data.frame} of\n",
       "bias-corrected log abundance table.\n",
       "\\item{}  \\code{ss\\_tab}, a \\code{data.frame} of sensitivity scores for\n",
       "pseudo-count addition to 0s.\n",
       "\\item{}  \\code{zero\\_ind}, a logical \\code{data.frame} with TRUE\n",
       "indicating the taxon is detected to contain structural zeros in\n",
       "some specific groups.\n",
       "\\item{}  \\code{samp\\_frac}, a numeric vector of estimated sampling\n",
       "fractions in log scale (natural log).\n",
       "\\item{}  \\code{delta\\_em}, estimated sample-specific biases\n",
       "through E-M algorithm.\n",
       "\\item{}  \\code{delta\\_wls}, estimated sample-specific biases through\n",
       "weighted least squares (WLS) algorithm.\n",
       "\\item{}  \\code{res},  a \\code{data.frame} containing ANCOM-BC2 primary\n",
       "result:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{}  columns started with \\code{lfc}: log fold changes\n",
       "obtained from the ANCOM-BC2 log-linear (natural log) model.\n",
       "\\item{}  columns started with \\code{se}: standard errors (SEs) of\n",
       "\\code{lfc}.\n",
       "\\item{}  columns started with \\code{W}: test statistics.\n",
       "\\code{W = lfc/se}.\n",
       "\\item{}  columns started with \\code{p}: p-values. P-values are\n",
       "obtained from two-sided Z-test using the test statistic \\code{W}.\n",
       "\\item{}  columns started with \\code{q}: adjusted p-values.\n",
       "Adjusted p-values are obtained by applying \\code{p\\_adj\\_method}\n",
       "to \\code{p}.\n",
       "\\item{}  columns started with \\code{diff}: TRUE if the\n",
       "taxon is significant (has \\code{q} less than \\code{alpha}).\n",
       "\\item{}  columns started with \\code{passed\\_ss}: TRUE if the\n",
       "taxon passed the sensitivity analysis, i.e., adding different\n",
       "pseudo-counts to 0s would not change the results.\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\\item{}  columns started with \\code{diff\\_robust}: TRUE if the\n",
       "taxon is significant (has \\code{q} less than \\code{alpha}) and robust\n",
       "in the sensitivity analysis (\\code{passed\\_ss} is \\code{TRUE}).\n",
       "\\item{}  \\code{res\\_global},  a \\code{data.frame} containing ANCOM-BC2\n",
       "global test result for the variable specified in \\code{group},\n",
       "each column is:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{}  \\code{W}, test statistics.\n",
       "\\item{}  \\code{p\\_val}, p-values, which are obtained from two-sided\n",
       "Chi-square test using \\code{W}.\n",
       "\\item{}  \\code{q\\_val}, adjusted p-values. Adjusted p-values are\n",
       "obtained by applying \\code{p\\_adj\\_method} to \\code{p\\_val}.\n",
       "\\item{}  \\code{diff\\_abn}, A logical vector. TRUE if the taxon has\n",
       "\\code{q\\_val} less than \\code{alpha}.\n",
       "\\item{}  \\code{passed\\_ss}, A logical vector. TRUE if the taxon has\n",
       "passed the sensitivity analysis.\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\\item{}  \\code{res\\_pair},  a \\code{data.frame} containing ANCOM-BC2\n",
       "pairwise directional test result for the variable specified in\n",
       "\\code{group}:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{}  columns started with \\code{lfc}: log fold changes.\n",
       "\\item{}  columns started with \\code{se}: standard errors (SEs).\n",
       "\\item{}  columns started with \\code{W}: test statistics.\n",
       "\\item{}  columns started with \\code{p}: p-values.\n",
       "\\item{}  columns started with \\code{q}: adjusted p-values.\n",
       "\\item{}  columns started with \\code{diff}: TRUE if the\n",
       "taxon is significant (has \\code{q} less than \\code{alpha}).\n",
       "\\item{}  columns started with \\code{passed\\_ss}: TRUE if the\n",
       "taxon has passed the sensitivity analysis.\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\\item{}  \\code{res\\_dunn},  a \\code{data.frame} containing ANCOM-BC2\n",
       "Dunnett's type of test result for the variable specified in\n",
       "\\code{group}:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{}  columns started with \\code{lfc}: log fold changes.\n",
       "\\item{}  columns started with \\code{se}: standard errors (SEs).\n",
       "\\item{}  columns started with \\code{W}: test statistics.\n",
       "\\item{}  columns started with \\code{p}: p-values.\n",
       "\\item{}  columns started with \\code{q}: adjusted p-values.\n",
       "\\item{}  columns started with \\code{diff}: TRUE if the\n",
       "taxon is significant (has \\code{q} less than \\code{alpha}).\n",
       "\\item{}  columns started with \\code{passed\\_ss}: TRUE if the\n",
       "taxon has passed the sensitivity analysis.\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\\item{}  \\code{res\\_trend},  a \\code{data.frame} containing ANCOM-BC2\n",
       "trend test result for the variable specified in\n",
       "\\code{group}:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{}  columns started with \\code{lfc}: log fold changes.\n",
       "\\item{}  columns started with \\code{se}: standard errors (SEs).\n",
       "\\item{}  \\code{W}: test statistics.\n",
       "\\item{}  \\code{p\\_val}: p-values.\n",
       "\\item{}  \\code{q\\_val}: adjusted p-values.\n",
       "\\item{}  \\code{diff\\_abn}: TRUE if the\n",
       "taxon is significant (has \\code{q} less than \\code{alpha}).\n",
       "\\item{}  \\code{passed\\_ss}, A logical vector. TRUE if the taxon has\n",
       "passed the sensitivity analysis.\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Author}\n",
       "Huang Lin\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{References}\n",
       "Kaul A, Mandal S, Davidov O, Peddada SD (2017).\n",
       "``Analysis of microbiome data in the presence of excess zeros.''\n",
       "\\emph{Frontiers in microbiology}, \\bold{8}, 2114.\n",
       "\n",
       "Lin H, Peddada SD (2020).\n",
       "``Analysis of compositions of microbiomes with bias correction.''\n",
       "\\emph{Nature communications}, \\bold{11}(1), 1--11.\n",
       "\n",
       "Tusher VG, Tibshirani R, Chu G (2001).\n",
       "``Significance analysis of microarrays applied to the ionizing radiation response.''\n",
       "\\emph{Proceedings of the National Academy of Sciences}, \\bold{98}(9), 5116--5121.\n",
       "\n",
       "Costea PI, Zeller G, Sunagawa S, Bork P (2014).\n",
       "``A fair comparison.''\n",
       "\\emph{Nature methods}, \\bold{11}(4), 359--359.\n",
       "\n",
       "Paulson JN, Bravo HC, Pop M (2014).\n",
       "``Reply to:\" a fair comparison\".''\n",
       "\\emph{Nature methods}, \\bold{11}(4), 359--360.\n",
       "\n",
       "Guo W, Sarkar SK, Peddada SD (2010).\n",
       "``Controlling false discoveries in multidimensional directional decisions, with applications to gene expression data on ordered categories.''\n",
       "\\emph{Biometrics}, \\bold{66}(2), 485--492.\n",
       "\n",
       "Grandhi A, Guo W, Peddada SD (2016).\n",
       "``A multiple testing procedure for multi-dimensional pairwise comparisons with application to gene expression studies.''\n",
       "\\emph{BMC bioinformatics}, \\bold{17}(1), 1--12.\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{SeeAlso}\n",
       "\\code{\\LinkA{ancom}{ancom}} \\code{\\LinkA{ancombc}{ancombc}}\n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "library(ANCOMBC)\n",
       "if (requireNamespace(\"microbiome\", quietly = TRUE)) {\n",
       "    data(dietswap, package = \"microbiome\")\n",
       "\n",
       "    # run ancombc2 function\n",
       "    set.seed(123)\n",
       "    # Note that setting max_iter = 1 and B = 1 is only for the sake of speed\n",
       "    # Use default or larger values for max_iter and B for better performance\n",
       "    out = ancombc2(data = dietswap, tax_level = \"Family\",\n",
       "                   fix_formula = \"nationality + timepoint + bmi_group\",\n",
       "                   rand_formula = NULL,\n",
       "                   p_adj_method = \"holm\", pseudo_sens = TRUE,\n",
       "                   prv_cut = 0.10, lib_cut = 1000, s0_perc = 0.05,\n",
       "                   group = \"bmi_group\", struc_zero = TRUE, neg_lb = TRUE,\n",
       "                   alpha = 0.05, n_cl = 1, verbose = TRUE,\n",
       "                   global = TRUE, pairwise = TRUE, dunnet = TRUE, trend = TRUE,\n",
       "                   iter_control = list(tol = 1e-2, max_iter = 1, verbose = TRUE),\n",
       "                   em_control = list(tol = 1e-5, max_iter = 1),\n",
       "                   lme_control = lme4::lmerControl(),\n",
       "                   mdfdr_control = list(fwer_ctrl_method = \"holm\", B = 1),\n",
       "                   trend_control = list(contrast =\n",
       "                                            list(matrix(c(1, 0, -1, 1),\n",
       "                                                        nrow = 2,\n",
       "                                                        byrow = TRUE)),\n",
       "                                        node = list(2),\n",
       "                                        solver = \"ECOS\",\n",
       "                                        B = 1))\n",
       "    res_prim = out$res\n",
       "    res_global = out$res_global\n",
       "    res_pair = out$res_pair\n",
       "    res_dunn = out$res_dunn\n",
       "    res_trend = out$res_trend\n",
       "} else {\n",
       "    message(\"The 'microbiome' package is not installed. Please install it to use this example.\")\n",
       "}\n",
       "\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "ancombc2                package:ANCOMBC                R Documentation\n",
       "\n",
       "_\bA_\bn_\ba_\bl_\by_\bs_\bi_\bs _\bo_\bf _\bC_\bo_\bm_\bp_\bo_\bs_\bi_\bt_\bi_\bo_\bn_\bs _\bo_\bf _\bM_\bi_\bc_\br_\bo_\bb_\bi_\bo_\bm_\be_\bs _\bw_\bi_\bt_\bh _\bB_\bi_\ba_\bs _\bC_\bo_\br_\br_\be_\bc_\bt_\bi_\bo_\bn _\b2\n",
       "(_\bA_\bN_\bC_\bO_\bM-_\bB_\bC_\b2)\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     Determine taxa whose absolute abundances, per unit volume, of the\n",
       "     ecosystem (e.g., gut) are significantly different with changes in\n",
       "     the covariate of interest (e.g., group). The current version of\n",
       "     ‘ancombc2’ function implements Analysis of Compositions of\n",
       "     Microbiomes with Bias Correction (ANCOM-BC2) in cross-sectional\n",
       "     and repeated measurements data. In addition to the two-group\n",
       "     comparison, ANCOM-BC2 also supports testing for continuous\n",
       "     covariates and multi-group comparisons, including the global test,\n",
       "     pairwise directional test, Dunnett's type of test, and trend test.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     ancombc2(\n",
       "       data,\n",
       "       taxa_are_rows = TRUE,\n",
       "       assay.type = assay_name,\n",
       "       assay_name = \"counts\",\n",
       "       rank = tax_level,\n",
       "       tax_level = NULL,\n",
       "       aggregate_data = NULL,\n",
       "       meta_data = NULL,\n",
       "       fix_formula,\n",
       "       rand_formula = NULL,\n",
       "       p_adj_method = \"holm\",\n",
       "       pseudo = 0,\n",
       "       pseudo_sens = TRUE,\n",
       "       prv_cut = 0.1,\n",
       "       lib_cut = 0,\n",
       "       s0_perc = 0.05,\n",
       "       group = NULL,\n",
       "       struc_zero = FALSE,\n",
       "       neg_lb = FALSE,\n",
       "       alpha = 0.05,\n",
       "       n_cl = 1,\n",
       "       verbose = TRUE,\n",
       "       global = FALSE,\n",
       "       pairwise = FALSE,\n",
       "       dunnet = FALSE,\n",
       "       trend = FALSE,\n",
       "       iter_control = list(tol = 0.01, max_iter = 20, verbose = FALSE),\n",
       "       em_control = list(tol = 1e-05, max_iter = 100),\n",
       "       lme_control = lme4::lmerControl(),\n",
       "       mdfdr_control = list(fwer_ctrl_method = \"holm\", B = 100),\n",
       "       trend_control = list(contrast = NULL, node = NULL, solver = \"ECOS\", B = 100)\n",
       "     )\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "    data: the input data. The ‘data’ parameter should be either a\n",
       "          ‘matrix’, ‘data.frame’, ‘phyloseq’ or a\n",
       "          ‘TreeSummarizedExperiment’ object. Both ‘phyloseq’ and\n",
       "          ‘TreeSummarizedExperiment’ objects consist of a feature table\n",
       "          (microbial count table), a sample metadata table, a taxonomy\n",
       "          table (optional), and a phylogenetic tree (optional). If a\n",
       "          ‘matrix’ or ‘data.frame’ is provided, ensure that the row\n",
       "          names of the ‘metadata’ match the sample names (column names\n",
       "          if ‘taxa_are_rows’ is TRUE, and row names otherwise) in\n",
       "          ‘data’. if a ‘phyloseq’ or a ‘TreeSummarizedExperiment’ is\n",
       "          used, this standard has already been enforced. For detailed\n",
       "          information, refer to ‘?phyloseq::phyloseq’ or\n",
       "          ‘?TreeSummarizedExperiment::TreeSummarizedExperiment’. It is\n",
       "          recommended to use low taxonomic levels, such as OTU or\n",
       "          species level, as the estimation of sampling fractions\n",
       "          requires a large number of taxa.\n",
       "\n",
       "taxa_are_rows: logical. Whether taxa are positioned in the rows of the\n",
       "          feature table. Default is TRUE.\n",
       "\n",
       "assay.type: alias for ‘assay_name’.\n",
       "\n",
       "assay_name: character. Name of the count table in the data object (only\n",
       "          applicable if data object is a ‘(Tree)SummarizedExperiment’).\n",
       "          Default is \"counts\". See ‘?SummarizedExperiment::assay’ for\n",
       "          more details.\n",
       "\n",
       "    rank: alias for ‘tax_level’.\n",
       "\n",
       "tax_level: character. The taxonomic or non taxonomic(rowData) level of\n",
       "          interest. The input data can be analyzed at any taxonomic or\n",
       "          rowData level without prior agglomeration. Note that\n",
       "          ‘tax_level’ must be a value from ‘taxonomyRanks’ or\n",
       "          ‘rowData’, which includes \"Kingdom\", \"Phylum\" \"Class\",\n",
       "          \"Order\", \"Family\" \"Genus\" \"Species\" etc. See\n",
       "          ‘?mia::taxonomyRanks’ for more details. Default is NULL,\n",
       "          i.e., do not perform agglomeration, and the ANCOM-BC2\n",
       "          analysis will be performed at the lowest taxonomic level of\n",
       "          the input ‘data’.\n",
       "\n",
       "aggregate_data: The abundance data that has been aggregated to the\n",
       "          desired taxonomic level. This parameter is required only when\n",
       "          the input data is in ‘matrix’ or ‘data.frame’ format. For\n",
       "          ‘phyloseq’ or ‘TreeSummarizedExperiment’ data, aggregation is\n",
       "          performed by specifying the ‘tax_level’ parameter.\n",
       "\n",
       "meta_data: a ‘data.frame’ containing sample metadata. This parameter is\n",
       "          mandatory when the input ‘data’ is a generic ‘matrix’ or\n",
       "          ‘data.frame’. Ensure that the row names of the ‘metadata’\n",
       "          match the sample names (column names if ‘taxa_are_rows’ is\n",
       "          TRUE, and row names otherwise) in ‘data’.\n",
       "\n",
       "fix_formula: the character string expresses how the microbial absolute\n",
       "          abundances for each taxon depend on the fixed effects in\n",
       "          metadata. When specifying the ‘fix_formula’, make sure to\n",
       "          include the ‘group’ variable in the formula if it is not\n",
       "          NULL.\n",
       "\n",
       "rand_formula: the character string expresses how the microbial absolute\n",
       "          abundances for each taxon depend on the random effects in\n",
       "          metadata. ANCOM-BC2 follows the ‘lmerTest’ package in\n",
       "          formulating the random effects. See ‘?lmerTest::lmer’ for\n",
       "          more details. Default is ‘NULL’.\n",
       "\n",
       "p_adj_method: character. method to adjust p-values. Default is \"holm\".\n",
       "          Options include \"holm\", \"hochberg\", \"hommel\", \"bonferroni\",\n",
       "          \"BH\", \"BY\", \"fdr\", \"none\". See ‘?stats::p.adjust’ for more\n",
       "          details.\n",
       "\n",
       "  pseudo: A small positive value (default: 1) added to all counts\n",
       "          before log transformation to avoid numerical issues caused by\n",
       "          log(0).\n",
       "\n",
       "pseudo_sens: logical. Whether to perform the sensitivity analysis to\n",
       "          the pseudo-count addition. Default is ‘TRUE’. While ANCOM-BC2\n",
       "          utilizes complete data (nonzero counts) by default for its\n",
       "          analysis, a comprehensive evaluation of result robustness is\n",
       "          performed by assessing how pseudo-count addition to zeros may\n",
       "          affect the outcomes. For a detailed discussion on this\n",
       "          sensitivity analysis, refer to the ‘Details’ section.\n",
       "\n",
       " prv_cut: a numerical fraction between 0 and 1. Taxa with prevalences\n",
       "          (the proportion of samples in which the taxon is present)\n",
       "          less than ‘prv_cut’ will be excluded in the analysis. For\n",
       "          example, if there are 100 samples, and a taxon has nonzero\n",
       "          counts present in less than 100*prv_cut samples, it will not\n",
       "          be considered in the analysis. Default is 0.10.\n",
       "\n",
       " lib_cut: a numerical threshold for filtering samples based on library\n",
       "          sizes. Samples with library sizes less than ‘lib_cut’ will be\n",
       "          excluded in the analysis. Default is 0, i.e. do not discard\n",
       "          any sample.\n",
       "\n",
       " s0_perc: a numerical fraction between 0 and 1. Inspired by\n",
       "          Significance Analysis of Microarrays (SAM) methodology, a\n",
       "          small positive constant is added to the denominator of\n",
       "          ANCOM-BC2 test statistic corresponding to each taxon to avoid\n",
       "          the significance due to extremely small standard errors,\n",
       "          especially for rare taxa. This small positive constant is\n",
       "          chosen as ‘s0_perc’-th percentile of standard error values\n",
       "          for each fixed effect. Default is 0.05 (5th percentile).\n",
       "\n",
       "   group: character. the name of the group variable in metadata. The\n",
       "          ‘group’ parameter should be a character string representing\n",
       "          the name of the group variable in the metadata. The ‘group’\n",
       "          variable should be discrete, meaning it consists of\n",
       "          categorical values. Specifying the ‘group’ variable is\n",
       "          required if you are interested in detecting structural zeros\n",
       "          and performing performing multi-group comparisons (global\n",
       "          test, pairwise directional test, Dunnett's type of test, and\n",
       "          trend test). However, if these analyses are not of interest\n",
       "          to you, you can leave the ‘group’ parameter as NULL. If the\n",
       "          ‘group’ variable of interest contains only two categories,\n",
       "          you can also leave the ‘group’ parameter as NULL. Default is\n",
       "          NULL.\n",
       "\n",
       "struc_zero: logical. Whether to detect structural zeros based on\n",
       "          ‘group’. Default is FALSE. See ‘Details’ for a more\n",
       "          comprehensive discussion on structural zeros.\n",
       "\n",
       "  neg_lb: logical. Whether to classify a taxon as a structural zero\n",
       "          using its asymptotic lower bound. Default is FALSE.\n",
       "\n",
       "   alpha: numeric. Level of significance. Default is 0.05.\n",
       "\n",
       "    n_cl: numeric. The number of nodes to be forked. For details, see\n",
       "          ‘?parallel::makeCluster’. Default is 1 (no parallel\n",
       "          computing).\n",
       "\n",
       " verbose: logical. Whether to generate verbose output during the\n",
       "          ANCOM-BC2 fitting process. Default is FALSE.\n",
       "\n",
       "  global: logical. Whether to perform the global test. Default is\n",
       "          FALSE.\n",
       "\n",
       "pairwise: logical. Whether to perform the pairwise directional test.\n",
       "          Default is FALSE.\n",
       "\n",
       "  dunnet: logical. Whether to perform the Dunnett's type of test.\n",
       "          Default is FALSE.\n",
       "\n",
       "   trend: logical. Whether to perform trend test. Default is FALSE.\n",
       "\n",
       "iter_control: a named list of control parameters for the iterative MLE\n",
       "          or RMEL algorithm, including 1) ‘tol’: the iteration\n",
       "          convergence tolerance (default is 1e-02), 2) ‘max_iter’: the\n",
       "          maximum number of iterations (default is 20), and\n",
       "          3)‘verbose’: whether to show the verbose output (default is\n",
       "          FALSE).\n",
       "\n",
       "em_control: a named list of control parameters for the E-M algorithm,\n",
       "          including 1) ‘tol’: the iteration convergence tolerance\n",
       "          (default is 1e-05) and 2) ‘max_iter’: the maximum number of\n",
       "          iterations (default is 100).\n",
       "\n",
       "lme_control: a list of control parameters for mixed model fitting. See\n",
       "          ‘?lme4::lmerControl’ for details.\n",
       "\n",
       "mdfdr_control: a named list of control parameters for mixed directional\n",
       "          false discover rate (mdFDR), including 1) ‘fwer_ctrl_method’:\n",
       "          family wise error (FWER) controlling procedure, such as\n",
       "          \"holm\", \"hochberg\", \"bonferroni\", etc (default is \"holm\") and\n",
       "          2) ‘B’: the number of bootstrap samples (default is 100).\n",
       "          Increase ‘B’ will lead to a more accurate p-values. See\n",
       "          ‘Details’ for a more comprehensive discussion on mdFDR.\n",
       "\n",
       "trend_control: a named list of control parameters for the trend test,\n",
       "          including 1) ‘contrast’: the list of contrast matrices for\n",
       "          constructing inequalities, 2) ‘node’: the list of positions\n",
       "          for the nodal parameter, 3) ‘solver’: a string indicating the\n",
       "          solver to use (default is \"ECOS\"), and 4) ‘B’: the number of\n",
       "          bootstrap samples (default is 100). Increase ‘B’ will lead to\n",
       "          a more accurate p-values. See ‘vignette’ for the\n",
       "          corresponding trend test examples.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     A taxon is considered to have structural zeros in some (>=1)\n",
       "     groups if it is completely (or nearly completely) missing in these\n",
       "     groups. For instance, suppose there are three groups: g1, g2, and\n",
       "     g3. If the counts of taxon A in g1 are 0 but nonzero in g2 and g3,\n",
       "     then taxon A will be considered to contain structural zeros in g1.\n",
       "     In this example, taxon A is declared to be differentially abundant\n",
       "     between g1 and g2, g1 and g3, and consequently, it is globally\n",
       "     differentially abundant with respect to this group variable. Such\n",
       "     taxa are not further analyzed using ANCOM-BC2, but the results are\n",
       "     summarized in the overall summary. For more details about the\n",
       "     structural zeros, please go to the ANCOM-II paper. Setting ‘neg_lb\n",
       "     = TRUE’ indicates that you are using both criteria stated in\n",
       "     section 3.2 of ANCOM-II to detect structural zeros; otherwise, the\n",
       "     algorithm will only use the equation 1 in section 3.2 for\n",
       "     declaring structural zeros. Generally, it is recommended to set\n",
       "     ‘neg_lb = TRUE’ when the sample size per group is relatively large\n",
       "     (e.g. > 30).\n",
       "\n",
       "     Like other differential abundance analysis methods, ANCOM-BC2\n",
       "     applies a log transformation to the observed counts. However, the\n",
       "     presence of zero counts poses a challenge, and researchers often\n",
       "     consider adding a pseudo-count before the log transformation.\n",
       "     However, it has been shown that the choice of pseudo-count can\n",
       "     impact the results and lead to an inflated false positive rate\n",
       "     (Costea et al. (2014); Paulson, Bravo, and Pop (2014)). To address\n",
       "     this issue, we conduct a sensitivity analysis to assess the impact\n",
       "     of different pseudo-counts on zero counts for each taxon. This\n",
       "     involves adding various pseudo-count values (0.1, 0.5, and 1) to\n",
       "     the feature table and re-running the ANCOM-BC2 algorithm. The\n",
       "     sensitivity score for each taxon is calculated as the proportion\n",
       "     of times that the q-value exceeds the specified significance level\n",
       "     (alpha). If all q-values consistently show significance or\n",
       "     nonsignificance across different pseudo-counts and are consistent\n",
       "     with the results obtained without adding pseudo-counts to zero\n",
       "     counts (using the default settings), then the taxon is considered\n",
       "     not sensitive to the pseudo-count addition.\n",
       "\n",
       "     When performning pairwise directional (or Dunnett's type of) test,\n",
       "     the mixed directional false discover rate (mdFDR) should be taken\n",
       "     into account. The mdFDR is the combination of false discovery rate\n",
       "     due to multiple testing, multiple pairwise comparisons, and\n",
       "     directional tests within each pairwise comparison. For example,\n",
       "     suppose we have five taxa and three experimental groups: g1, g2,\n",
       "     and g3. Thus, we are performing five tests corresponding to five\n",
       "     taxa. For each taxon, we are also conducting three pairwise\n",
       "     comparisons (g1 vs. g2, g2 vs. g3, and g1 vs. g3). Within each\n",
       "     pairwise comparison, we wish to determine if the abundance has\n",
       "     increased or decreased or did not change (direction of the effect\n",
       "     size). Errors could occur in each step. The overall false\n",
       "     discovery rate is controlled by the mdFDR methodology we adopted\n",
       "     from Guo, Sarkar, and Peddada (2010) and Grandhi, Guo, and Peddada\n",
       "     (2016).\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     a ‘list’ with components:\n",
       "\n",
       "        • ‘feature_table’, a ‘data.frame’ of pre-processed (based on\n",
       "          ‘prv_cut’ and ‘lib_cut’) microbial count table.\n",
       "\n",
       "        • ‘bias_correct_log_table’, a ‘data.frame’ of bias-corrected\n",
       "          log abundance table.\n",
       "\n",
       "        • ‘ss_tab’, a ‘data.frame’ of sensitivity scores for\n",
       "          pseudo-count addition to 0s.\n",
       "\n",
       "        • ‘zero_ind’, a logical ‘data.frame’ with TRUE indicating the\n",
       "          taxon is detected to contain structural zeros in some\n",
       "          specific groups.\n",
       "\n",
       "        • ‘samp_frac’, a numeric vector of estimated sampling fractions\n",
       "          in log scale (natural log).\n",
       "\n",
       "        • ‘delta_em’, estimated sample-specific biases through E-M\n",
       "          algorithm.\n",
       "\n",
       "        • ‘delta_wls’, estimated sample-specific biases through\n",
       "          weighted least squares (WLS) algorithm.\n",
       "\n",
       "        • ‘res’, a ‘data.frame’ containing ANCOM-BC2 primary result:\n",
       "\n",
       "            • columns started with ‘lfc’: log fold changes obtained\n",
       "              from the ANCOM-BC2 log-linear (natural log) model.\n",
       "\n",
       "            • columns started with ‘se’: standard errors (SEs) of\n",
       "              ‘lfc’.\n",
       "\n",
       "            • columns started with ‘W’: test statistics.  ‘W = lfc/se’.\n",
       "\n",
       "            • columns started with ‘p’: p-values. P-values are obtained\n",
       "              from two-sided Z-test using the test statistic ‘W’.\n",
       "\n",
       "            • columns started with ‘q’: adjusted p-values.  Adjusted\n",
       "              p-values are obtained by applying ‘p_adj_method’ to ‘p’.\n",
       "\n",
       "            • columns started with ‘diff’: TRUE if the taxon is\n",
       "              significant (has ‘q’ less than ‘alpha’).\n",
       "\n",
       "            • columns started with ‘passed_ss’: TRUE if the taxon\n",
       "              passed the sensitivity analysis, i.e., adding different\n",
       "              pseudo-counts to 0s would not change the results.\n",
       "\n",
       "        • columns started with ‘diff_robust’: TRUE if the taxon is\n",
       "          significant (has ‘q’ less than ‘alpha’) and robust in the\n",
       "          sensitivity analysis (‘passed_ss’ is ‘TRUE’).\n",
       "\n",
       "        • ‘res_global’, a ‘data.frame’ containing ANCOM-BC2 global test\n",
       "          result for the variable specified in ‘group’, each column is:\n",
       "\n",
       "            • ‘W’, test statistics.\n",
       "\n",
       "            • ‘p_val’, p-values, which are obtained from two-sided\n",
       "              Chi-square test using ‘W’.\n",
       "\n",
       "            • ‘q_val’, adjusted p-values. Adjusted p-values are\n",
       "              obtained by applying ‘p_adj_method’ to ‘p_val’.\n",
       "\n",
       "            • ‘diff_abn’, A logical vector. TRUE if the taxon has\n",
       "              ‘q_val’ less than ‘alpha’.\n",
       "\n",
       "            • ‘passed_ss’, A logical vector. TRUE if the taxon has\n",
       "              passed the sensitivity analysis.\n",
       "\n",
       "        • ‘res_pair’, a ‘data.frame’ containing ANCOM-BC2 pairwise\n",
       "          directional test result for the variable specified in\n",
       "          ‘group’:\n",
       "\n",
       "            • columns started with ‘lfc’: log fold changes.\n",
       "\n",
       "            • columns started with ‘se’: standard errors (SEs).\n",
       "\n",
       "            • columns started with ‘W’: test statistics.\n",
       "\n",
       "            • columns started with ‘p’: p-values.\n",
       "\n",
       "            • columns started with ‘q’: adjusted p-values.\n",
       "\n",
       "            • columns started with ‘diff’: TRUE if the taxon is\n",
       "              significant (has ‘q’ less than ‘alpha’).\n",
       "\n",
       "            • columns started with ‘passed_ss’: TRUE if the taxon has\n",
       "              passed the sensitivity analysis.\n",
       "\n",
       "        • ‘res_dunn’, a ‘data.frame’ containing ANCOM-BC2 Dunnett's\n",
       "          type of test result for the variable specified in ‘group’:\n",
       "\n",
       "            • columns started with ‘lfc’: log fold changes.\n",
       "\n",
       "            • columns started with ‘se’: standard errors (SEs).\n",
       "\n",
       "            • columns started with ‘W’: test statistics.\n",
       "\n",
       "            • columns started with ‘p’: p-values.\n",
       "\n",
       "            • columns started with ‘q’: adjusted p-values.\n",
       "\n",
       "            • columns started with ‘diff’: TRUE if the taxon is\n",
       "              significant (has ‘q’ less than ‘alpha’).\n",
       "\n",
       "            • columns started with ‘passed_ss’: TRUE if the taxon has\n",
       "              passed the sensitivity analysis.\n",
       "\n",
       "        • ‘res_trend’, a ‘data.frame’ containing ANCOM-BC2 trend test\n",
       "          result for the variable specified in ‘group’:\n",
       "\n",
       "            • columns started with ‘lfc’: log fold changes.\n",
       "\n",
       "            • columns started with ‘se’: standard errors (SEs).\n",
       "\n",
       "            • ‘W’: test statistics.\n",
       "\n",
       "            • ‘p_val’: p-values.\n",
       "\n",
       "            • ‘q_val’: adjusted p-values.\n",
       "\n",
       "            • ‘diff_abn’: TRUE if the taxon is significant (has ‘q’\n",
       "              less than ‘alpha’).\n",
       "\n",
       "            • ‘passed_ss’, A logical vector. TRUE if the taxon has\n",
       "              passed the sensitivity analysis.\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     Huang Lin\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     Kaul A, Mandal S, Davidov O, Peddada SD (2017). “Analysis of\n",
       "     microbiome data in the presence of excess zeros.” _Frontiers in\n",
       "     microbiology_, *8*, 2114.\n",
       "\n",
       "     Lin H, Peddada SD (2020). “Analysis of compositions of microbiomes\n",
       "     with bias correction.” _Nature communications_, *11*(1), 1-11.\n",
       "\n",
       "     Tusher VG, Tibshirani R, Chu G (2001). “Significance analysis of\n",
       "     microarrays applied to the ionizing radiation response.”\n",
       "     _Proceedings of the National Academy of Sciences_, *98*(9),\n",
       "     5116-5121.\n",
       "\n",
       "     Costea PI, Zeller G, Sunagawa S, Bork P (2014). “A fair\n",
       "     comparison.” _Nature methods_, *11*(4), 359-359.\n",
       "\n",
       "     Paulson JN, Bravo HC, Pop M (2014). “Reply to:\" a fair\n",
       "     comparison\".” _Nature methods_, *11*(4), 359-360.\n",
       "\n",
       "     Guo W, Sarkar SK, Peddada SD (2010). “Controlling false\n",
       "     discoveries in multidimensional directional decisions, with\n",
       "     applications to gene expression data on ordered categories.”\n",
       "     _Biometrics_, *66*(2), 485-492.\n",
       "\n",
       "     Grandhi A, Guo W, Peddada SD (2016). “A multiple testing procedure\n",
       "     for multi-dimensional pairwise comparisons with application to\n",
       "     gene expression studies.” _BMC bioinformatics_, *17*(1), 1-12.\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     ‘ancom’ ‘ancombc’\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     library(ANCOMBC)\n",
       "     if (requireNamespace(\"microbiome\", quietly = TRUE)) {\n",
       "         data(dietswap, package = \"microbiome\")\n",
       "     \n",
       "         # run ancombc2 function\n",
       "         set.seed(123)\n",
       "         # Note that setting max_iter = 1 and B = 1 is only for the sake of speed\n",
       "         # Use default or larger values for max_iter and B for better performance\n",
       "         out = ancombc2(data = dietswap, tax_level = \"Family\",\n",
       "                        fix_formula = \"nationality + timepoint + bmi_group\",\n",
       "                        rand_formula = NULL,\n",
       "                        p_adj_method = \"holm\", pseudo_sens = TRUE,\n",
       "                        prv_cut = 0.10, lib_cut = 1000, s0_perc = 0.05,\n",
       "                        group = \"bmi_group\", struc_zero = TRUE, neg_lb = TRUE,\n",
       "                        alpha = 0.05, n_cl = 1, verbose = TRUE,\n",
       "                        global = TRUE, pairwise = TRUE, dunnet = TRUE, trend = TRUE,\n",
       "                        iter_control = list(tol = 1e-2, max_iter = 1, verbose = TRUE),\n",
       "                        em_control = list(tol = 1e-5, max_iter = 1),\n",
       "                        lme_control = lme4::lmerControl(),\n",
       "                        mdfdr_control = list(fwer_ctrl_method = \"holm\", B = 1),\n",
       "                        trend_control = list(contrast =\n",
       "                                                 list(matrix(c(1, 0, -1, 1),\n",
       "                                                             nrow = 2,\n",
       "                                                             byrow = TRUE)),\n",
       "                                             node = list(2),\n",
       "                                             solver = \"ECOS\",\n",
       "                                             B = 1))\n",
       "         res_prim = out$res\n",
       "         res_global = out$res_global\n",
       "         res_pair = out$res_pair\n",
       "         res_dunn = out$res_dunn\n",
       "         res_trend = out$res_trend\n",
       "     } else {\n",
       "         message(\"The 'microbiome' package is not installed. Please install it to use this example.\")\n",
       "     }\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?ancombc2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4795eecc-4042-4996-ab9e-ebd9d7cf1c63",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # creating 1 df\n",
    "\n",
    "# baseline_saliva_df$Sample = 'Saliva'\n",
    "# baseline_stool_df$Sample = 'Stool'\n",
    "\n",
    "# combined_df = full_join(baseline_saliva_df, baseline_stool_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d471bc4c-3c08-46a2-840b-ee98b86fe5a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 37 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Original_Sample_ID</th><th scope=col>Subject_No</th><th scope=col>Time_Point</th><th scope=col>Treatment</th><th scope=col>Outcome</th><th scope=col>Age</th><th scope=col>IgE</th><th scope=col>IgE_Binarized</th><th scope=col>k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Prevotellaceae.g__Alloprevotella.s__sp._HMT_308</th><th scope=col>k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Porphyromonadaceae.g__Porphyromonas.__</th><th scope=col>k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.s__sp._HMT_221</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>Pre_1</th><td>FS01baseline</td><td> 1</td><td>Pre</td><td>P</td><td>F </td><td>3.838356</td><td> 41.00</td><td>High</td><td>  5</td><td>   0</td><td>  23</td></tr>\n",
       "\t<tr><th scope=row>Pre_2</th><td>FS02baseline</td><td> 2</td><td>Pre</td><td>A</td><td>F </td><td>1.873973</td><td> 14.20</td><td>Low </td><td> 78</td><td>   0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_3</th><td>FS04baseline</td><td> 3</td><td>Pre</td><td>P</td><td>F </td><td>3.715068</td><td> 67.40</td><td>High</td><td>  0</td><td>   0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_4</th><td>FS05baseline</td><td> 4</td><td>Pre</td><td>A</td><td>F </td><td>3.991781</td><td>  9.59</td><td>Low </td><td>  0</td><td>  13</td><td>   9</td></tr>\n",
       "\t<tr><th scope=row>Pre_5</th><td>FS06baseline</td><td> 5</td><td>Pre</td><td>P</td><td>F </td><td>2.356164</td><td>106.00</td><td>High</td><td>  0</td><td>  32</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_6</th><td>FS07baseline</td><td> 6</td><td>Pre</td><td>A</td><td>SU</td><td>1.419178</td><td>  1.32</td><td>Low </td><td>  0</td><td> 387</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_8</th><td>FS09baseline</td><td> 8</td><td>Pre</td><td>P</td><td>F </td><td>1.112329</td><td> 19.10</td><td>High</td><td>  0</td><td>   0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_9</th><td>FS10baseline</td><td> 9</td><td>Pre</td><td>A</td><td>SU</td><td>1.389041</td><td>  0.42</td><td>Low </td><td>  0</td><td> 703</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_10</th><td>FS11baseline</td><td>10</td><td>Pre</td><td>P</td><td>F </td><td>2.191781</td><td>  1.74</td><td>Low </td><td>  0</td><td>   0</td><td>   9</td></tr>\n",
       "\t<tr><th scope=row>Pre_11</th><td>FS12baseline</td><td>11</td><td>Pre</td><td>A</td><td>F </td><td>1.567123</td><td>  4.34</td><td>Low </td><td>  0</td><td> 215</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_12</th><td>FS13baseline</td><td>12</td><td>Pre</td><td>P</td><td>F </td><td>1.032877</td><td>    NA</td><td>NA  </td><td>  0</td><td>  38</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_13</th><td>FS14baseline</td><td>13</td><td>Pre</td><td>P</td><td>F </td><td>      NA</td><td>    NA</td><td>Low </td><td>  0</td><td>  44</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_14</th><td>FS15baseline</td><td>14</td><td>Pre</td><td>A</td><td>F </td><td>3.912329</td><td>105.00</td><td>High</td><td>327</td><td>   0</td><td> 575</td></tr>\n",
       "\t<tr><th scope=row>Pre_15</th><td>FS16baseline</td><td>15</td><td>Pre</td><td>P</td><td>F </td><td>3.641096</td><td>  3.70</td><td>Low </td><td>  0</td><td>  93</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_16</th><td>FS17baseline</td><td>16</td><td>Pre</td><td>A</td><td>F </td><td>2.934247</td><td>  6.23</td><td>Low </td><td> 64</td><td>  75</td><td>  13</td></tr>\n",
       "\t<tr><th scope=row>Pre_17</th><td>FS18baseline</td><td>17</td><td>Pre</td><td>P</td><td>F </td><td>2.767123</td><td> 18.90</td><td>High</td><td> 42</td><td>  88</td><td> 115</td></tr>\n",
       "\t<tr><th scope=row>Pre_18</th><td>FS19baseline</td><td>18</td><td>Pre</td><td>A</td><td>SU</td><td>2.221918</td><td>  4.30</td><td>Low </td><td>  0</td><td>   0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_19</th><td>FS20baseline</td><td>19</td><td>Pre</td><td>A</td><td>SU</td><td>3.638356</td><td>  0.31</td><td>Low </td><td>  0</td><td> 210</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_22</th><td>FS23baseline</td><td>22</td><td>Pre</td><td>P</td><td>F </td><td>2.495890</td><td>  4.57</td><td>Low </td><td>  0</td><td> 911</td><td> 117</td></tr>\n",
       "\t<tr><th scope=row>Pre_23</th><td>FS24baseline</td><td>23</td><td>Pre</td><td>A</td><td>SU</td><td>3.871233</td><td>  1.85</td><td>Low </td><td>  0</td><td>   0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_25</th><td>FS26baseline</td><td>25</td><td>Pre</td><td>P</td><td>F </td><td>3.356164</td><td> 98.70</td><td>High</td><td>  0</td><td>   8</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_27</th><td>FS28baseline</td><td>27</td><td>Pre</td><td>A</td><td>SU</td><td>3.553425</td><td>  0.83</td><td>Low </td><td>  8</td><td> 151</td><td>  59</td></tr>\n",
       "\t<tr><th scope=row>Pre_28</th><td>FS29baseline</td><td>28</td><td>Pre</td><td>A</td><td>F </td><td>3.643836</td><td> 13.70</td><td>Low </td><td>196</td><td>  37</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_29</th><td>FS30baseline</td><td>29</td><td>Pre</td><td>A</td><td>F </td><td>3.528767</td><td>180.00</td><td>High</td><td> 42</td><td> 107</td><td>   4</td></tr>\n",
       "\t<tr><th scope=row>Pre_30</th><td>FS31baseline</td><td>30</td><td>Pre</td><td>P</td><td>F </td><td>1.890411</td><td> 96.70</td><td>High</td><td>  0</td><td>   0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_31</th><td>FS32baseline</td><td>31</td><td>Pre</td><td>P</td><td>F </td><td>1.663014</td><td>113.00</td><td>High</td><td>860</td><td>  88</td><td> 360</td></tr>\n",
       "\t<tr><th scope=row>Pre_32</th><td>FS33baseline</td><td>32</td><td>Pre</td><td>A</td><td>SU</td><td>1.663014</td><td>113.00</td><td>Low </td><td>  0</td><td>2037</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_34</th><td>FS35baseline</td><td>34</td><td>Pre</td><td>A</td><td>F </td><td>2.972603</td><td>156.00</td><td>High</td><td>  0</td><td> 158</td><td>  14</td></tr>\n",
       "\t<tr><th scope=row>Pre_35</th><td>FS36baseline</td><td>35</td><td>Pre</td><td>P</td><td>F </td><td>3.030137</td><td> 14.10</td><td>Low </td><td> 67</td><td> 145</td><td>  86</td></tr>\n",
       "\t<tr><th scope=row>Pre_36</th><td>FS37baseline</td><td>36</td><td>Pre</td><td>A</td><td>F </td><td>1.290411</td><td>  8.98</td><td>Low </td><td>  0</td><td> 173</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_37</th><td>FS39baseline</td><td>37</td><td>Pre</td><td>P</td><td>F </td><td>1.715068</td><td> 19.90</td><td>High</td><td>  0</td><td> 655</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_39</th><td>FS41baseline</td><td>39</td><td>Pre</td><td>A</td><td>SU</td><td>1.145205</td><td>  5.18</td><td>Low </td><td>  0</td><td> 268</td><td> 116</td></tr>\n",
       "\t<tr><th scope=row>Pre_41</th><td>FS43baseline</td><td>41</td><td>Pre</td><td>A</td><td>SU</td><td>2.942466</td><td>  5.61</td><td>Low </td><td>  0</td><td> 357</td><td>1322</td></tr>\n",
       "\t<tr><th scope=row>Pre_42</th><td>FS44baseline</td><td>42</td><td>Pre</td><td>P</td><td>F </td><td>1.635616</td><td>  4.10</td><td>Low </td><td>  0</td><td> 909</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_43</th><td>FS45baseline</td><td>43</td><td>Pre</td><td>P</td><td>F </td><td>3.939726</td><td> 71.70</td><td>High</td><td>  0</td><td>   0</td><td>   0</td></tr>\n",
       "\t<tr><th scope=row>Pre_44</th><td>FS46baseline</td><td>44</td><td>Pre</td><td>A</td><td>SU</td><td>1.769863</td><td>  2.32</td><td>Low </td><td>  0</td><td> 617</td><td>2023</td></tr>\n",
       "\t<tr><th scope=row>Pre_45</th><td>FS47baseline</td><td>45</td><td>Pre</td><td>P</td><td>F </td><td>1.769863</td><td>  2.32</td><td>High</td><td>  0</td><td> 179</td><td>   0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 37 × 11\n",
       "\\begin{tabular}{r|lllllllllll}\n",
       "  & Original\\_Sample\\_ID & Subject\\_No & Time\\_Point & Treatment & Outcome & Age & IgE & IgE\\_Binarized & k\\_\\_Bacteria.p\\_\\_Bacteroidetes.c\\_\\_Bacteroidia.o\\_\\_Bacteroidales.f\\_\\_Prevotellaceae.g\\_\\_Alloprevotella.s\\_\\_sp.\\_HMT\\_308 & k\\_\\_Bacteria.p\\_\\_Bacteroidetes.c\\_\\_Bacteroidia.o\\_\\_Bacteroidales.f\\_\\_Porphyromonadaceae.g\\_\\_Porphyromonas.\\_\\_ & k\\_\\_Bacteria.p\\_\\_Fusobacteria.c\\_\\_Fusobacteriia.o\\_\\_Fusobacteriales.f\\_\\_Leptotrichiaceae.g\\_\\_Pseudoleptotrichia.s\\_\\_sp.\\_HMT\\_221\\\\\n",
       "  & <chr> & <dbl> & <chr> & <chr> & <fct> & <dbl> & <dbl> & <chr> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tPre\\_1 & FS01baseline &  1 & Pre & P & F  & 3.838356 &  41.00 & High &   5 &    0 &   23\\\\\n",
       "\tPre\\_2 & FS02baseline &  2 & Pre & A & F  & 1.873973 &  14.20 & Low  &  78 &    0 &    0\\\\\n",
       "\tPre\\_3 & FS04baseline &  3 & Pre & P & F  & 3.715068 &  67.40 & High &   0 &    0 &    0\\\\\n",
       "\tPre\\_4 & FS05baseline &  4 & Pre & A & F  & 3.991781 &   9.59 & Low  &   0 &   13 &    9\\\\\n",
       "\tPre\\_5 & FS06baseline &  5 & Pre & P & F  & 2.356164 & 106.00 & High &   0 &   32 &    0\\\\\n",
       "\tPre\\_6 & FS07baseline &  6 & Pre & A & SU & 1.419178 &   1.32 & Low  &   0 &  387 &    0\\\\\n",
       "\tPre\\_8 & FS09baseline &  8 & Pre & P & F  & 1.112329 &  19.10 & High &   0 &    0 &    0\\\\\n",
       "\tPre\\_9 & FS10baseline &  9 & Pre & A & SU & 1.389041 &   0.42 & Low  &   0 &  703 &    0\\\\\n",
       "\tPre\\_10 & FS11baseline & 10 & Pre & P & F  & 2.191781 &   1.74 & Low  &   0 &    0 &    9\\\\\n",
       "\tPre\\_11 & FS12baseline & 11 & Pre & A & F  & 1.567123 &   4.34 & Low  &   0 &  215 &    0\\\\\n",
       "\tPre\\_12 & FS13baseline & 12 & Pre & P & F  & 1.032877 &     NA & NA   &   0 &   38 &    0\\\\\n",
       "\tPre\\_13 & FS14baseline & 13 & Pre & P & F  &       NA &     NA & Low  &   0 &   44 &    0\\\\\n",
       "\tPre\\_14 & FS15baseline & 14 & Pre & A & F  & 3.912329 & 105.00 & High & 327 &    0 &  575\\\\\n",
       "\tPre\\_15 & FS16baseline & 15 & Pre & P & F  & 3.641096 &   3.70 & Low  &   0 &   93 &    0\\\\\n",
       "\tPre\\_16 & FS17baseline & 16 & Pre & A & F  & 2.934247 &   6.23 & Low  &  64 &   75 &   13\\\\\n",
       "\tPre\\_17 & FS18baseline & 17 & Pre & P & F  & 2.767123 &  18.90 & High &  42 &   88 &  115\\\\\n",
       "\tPre\\_18 & FS19baseline & 18 & Pre & A & SU & 2.221918 &   4.30 & Low  &   0 &    0 &    0\\\\\n",
       "\tPre\\_19 & FS20baseline & 19 & Pre & A & SU & 3.638356 &   0.31 & Low  &   0 &  210 &    0\\\\\n",
       "\tPre\\_22 & FS23baseline & 22 & Pre & P & F  & 2.495890 &   4.57 & Low  &   0 &  911 &  117\\\\\n",
       "\tPre\\_23 & FS24baseline & 23 & Pre & A & SU & 3.871233 &   1.85 & Low  &   0 &    0 &    0\\\\\n",
       "\tPre\\_25 & FS26baseline & 25 & Pre & P & F  & 3.356164 &  98.70 & High &   0 &    8 &    0\\\\\n",
       "\tPre\\_27 & FS28baseline & 27 & Pre & A & SU & 3.553425 &   0.83 & Low  &   8 &  151 &   59\\\\\n",
       "\tPre\\_28 & FS29baseline & 28 & Pre & A & F  & 3.643836 &  13.70 & Low  & 196 &   37 &    0\\\\\n",
       "\tPre\\_29 & FS30baseline & 29 & Pre & A & F  & 3.528767 & 180.00 & High &  42 &  107 &    4\\\\\n",
       "\tPre\\_30 & FS31baseline & 30 & Pre & P & F  & 1.890411 &  96.70 & High &   0 &    0 &    0\\\\\n",
       "\tPre\\_31 & FS32baseline & 31 & Pre & P & F  & 1.663014 & 113.00 & High & 860 &   88 &  360\\\\\n",
       "\tPre\\_32 & FS33baseline & 32 & Pre & A & SU & 1.663014 & 113.00 & Low  &   0 & 2037 &    0\\\\\n",
       "\tPre\\_34 & FS35baseline & 34 & Pre & A & F  & 2.972603 & 156.00 & High &   0 &  158 &   14\\\\\n",
       "\tPre\\_35 & FS36baseline & 35 & Pre & P & F  & 3.030137 &  14.10 & Low  &  67 &  145 &   86\\\\\n",
       "\tPre\\_36 & FS37baseline & 36 & Pre & A & F  & 1.290411 &   8.98 & Low  &   0 &  173 &    0\\\\\n",
       "\tPre\\_37 & FS39baseline & 37 & Pre & P & F  & 1.715068 &  19.90 & High &   0 &  655 &    0\\\\\n",
       "\tPre\\_39 & FS41baseline & 39 & Pre & A & SU & 1.145205 &   5.18 & Low  &   0 &  268 &  116\\\\\n",
       "\tPre\\_41 & FS43baseline & 41 & Pre & A & SU & 2.942466 &   5.61 & Low  &   0 &  357 & 1322\\\\\n",
       "\tPre\\_42 & FS44baseline & 42 & Pre & P & F  & 1.635616 &   4.10 & Low  &   0 &  909 &    0\\\\\n",
       "\tPre\\_43 & FS45baseline & 43 & Pre & P & F  & 3.939726 &  71.70 & High &   0 &    0 &    0\\\\\n",
       "\tPre\\_44 & FS46baseline & 44 & Pre & A & SU & 1.769863 &   2.32 & Low  &   0 &  617 & 2023\\\\\n",
       "\tPre\\_45 & FS47baseline & 45 & Pre & P & F  & 1.769863 &   2.32 & High &   0 &  179 &    0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 37 × 11\n",
       "\n",
       "| <!--/--> | Original_Sample_ID &lt;chr&gt; | Subject_No &lt;dbl&gt; | Time_Point &lt;chr&gt; | Treatment &lt;chr&gt; | Outcome &lt;fct&gt; | Age &lt;dbl&gt; | IgE &lt;dbl&gt; | IgE_Binarized &lt;chr&gt; | k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Prevotellaceae.g__Alloprevotella.s__sp._HMT_308 &lt;dbl&gt; | k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Porphyromonadaceae.g__Porphyromonas.__ &lt;dbl&gt; | k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.s__sp._HMT_221 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Pre_1 | FS01baseline |  1 | Pre | P | F  | 3.838356 |  41.00 | High |   5 |    0 |   23 |\n",
       "| Pre_2 | FS02baseline |  2 | Pre | A | F  | 1.873973 |  14.20 | Low  |  78 |    0 |    0 |\n",
       "| Pre_3 | FS04baseline |  3 | Pre | P | F  | 3.715068 |  67.40 | High |   0 |    0 |    0 |\n",
       "| Pre_4 | FS05baseline |  4 | Pre | A | F  | 3.991781 |   9.59 | Low  |   0 |   13 |    9 |\n",
       "| Pre_5 | FS06baseline |  5 | Pre | P | F  | 2.356164 | 106.00 | High |   0 |   32 |    0 |\n",
       "| Pre_6 | FS07baseline |  6 | Pre | A | SU | 1.419178 |   1.32 | Low  |   0 |  387 |    0 |\n",
       "| Pre_8 | FS09baseline |  8 | Pre | P | F  | 1.112329 |  19.10 | High |   0 |    0 |    0 |\n",
       "| Pre_9 | FS10baseline |  9 | Pre | A | SU | 1.389041 |   0.42 | Low  |   0 |  703 |    0 |\n",
       "| Pre_10 | FS11baseline | 10 | Pre | P | F  | 2.191781 |   1.74 | Low  |   0 |    0 |    9 |\n",
       "| Pre_11 | FS12baseline | 11 | Pre | A | F  | 1.567123 |   4.34 | Low  |   0 |  215 |    0 |\n",
       "| Pre_12 | FS13baseline | 12 | Pre | P | F  | 1.032877 |     NA | NA   |   0 |   38 |    0 |\n",
       "| Pre_13 | FS14baseline | 13 | Pre | P | F  |       NA |     NA | Low  |   0 |   44 |    0 |\n",
       "| Pre_14 | FS15baseline | 14 | Pre | A | F  | 3.912329 | 105.00 | High | 327 |    0 |  575 |\n",
       "| Pre_15 | FS16baseline | 15 | Pre | P | F  | 3.641096 |   3.70 | Low  |   0 |   93 |    0 |\n",
       "| Pre_16 | FS17baseline | 16 | Pre | A | F  | 2.934247 |   6.23 | Low  |  64 |   75 |   13 |\n",
       "| Pre_17 | FS18baseline | 17 | Pre | P | F  | 2.767123 |  18.90 | High |  42 |   88 |  115 |\n",
       "| Pre_18 | FS19baseline | 18 | Pre | A | SU | 2.221918 |   4.30 | Low  |   0 |    0 |    0 |\n",
       "| Pre_19 | FS20baseline | 19 | Pre | A | SU | 3.638356 |   0.31 | Low  |   0 |  210 |    0 |\n",
       "| Pre_22 | FS23baseline | 22 | Pre | P | F  | 2.495890 |   4.57 | Low  |   0 |  911 |  117 |\n",
       "| Pre_23 | FS24baseline | 23 | Pre | A | SU | 3.871233 |   1.85 | Low  |   0 |    0 |    0 |\n",
       "| Pre_25 | FS26baseline | 25 | Pre | P | F  | 3.356164 |  98.70 | High |   0 |    8 |    0 |\n",
       "| Pre_27 | FS28baseline | 27 | Pre | A | SU | 3.553425 |   0.83 | Low  |   8 |  151 |   59 |\n",
       "| Pre_28 | FS29baseline | 28 | Pre | A | F  | 3.643836 |  13.70 | Low  | 196 |   37 |    0 |\n",
       "| Pre_29 | FS30baseline | 29 | Pre | A | F  | 3.528767 | 180.00 | High |  42 |  107 |    4 |\n",
       "| Pre_30 | FS31baseline | 30 | Pre | P | F  | 1.890411 |  96.70 | High |   0 |    0 |    0 |\n",
       "| Pre_31 | FS32baseline | 31 | Pre | P | F  | 1.663014 | 113.00 | High | 860 |   88 |  360 |\n",
       "| Pre_32 | FS33baseline | 32 | Pre | A | SU | 1.663014 | 113.00 | Low  |   0 | 2037 |    0 |\n",
       "| Pre_34 | FS35baseline | 34 | Pre | A | F  | 2.972603 | 156.00 | High |   0 |  158 |   14 |\n",
       "| Pre_35 | FS36baseline | 35 | Pre | P | F  | 3.030137 |  14.10 | Low  |  67 |  145 |   86 |\n",
       "| Pre_36 | FS37baseline | 36 | Pre | A | F  | 1.290411 |   8.98 | Low  |   0 |  173 |    0 |\n",
       "| Pre_37 | FS39baseline | 37 | Pre | P | F  | 1.715068 |  19.90 | High |   0 |  655 |    0 |\n",
       "| Pre_39 | FS41baseline | 39 | Pre | A | SU | 1.145205 |   5.18 | Low  |   0 |  268 |  116 |\n",
       "| Pre_41 | FS43baseline | 41 | Pre | A | SU | 2.942466 |   5.61 | Low  |   0 |  357 | 1322 |\n",
       "| Pre_42 | FS44baseline | 42 | Pre | P | F  | 1.635616 |   4.10 | Low  |   0 |  909 |    0 |\n",
       "| Pre_43 | FS45baseline | 43 | Pre | P | F  | 3.939726 |  71.70 | High |   0 |    0 |    0 |\n",
       "| Pre_44 | FS46baseline | 44 | Pre | A | SU | 1.769863 |   2.32 | Low  |   0 |  617 | 2023 |\n",
       "| Pre_45 | FS47baseline | 45 | Pre | P | F  | 1.769863 |   2.32 | High |   0 |  179 |    0 |\n",
       "\n"
      ],
      "text/plain": [
       "       Original_Sample_ID Subject_No Time_Point Treatment Outcome Age     \n",
       "Pre_1  FS01baseline        1         Pre        P         F       3.838356\n",
       "Pre_2  FS02baseline        2         Pre        A         F       1.873973\n",
       "Pre_3  FS04baseline        3         Pre        P         F       3.715068\n",
       "Pre_4  FS05baseline        4         Pre        A         F       3.991781\n",
       "Pre_5  FS06baseline        5         Pre        P         F       2.356164\n",
       "Pre_6  FS07baseline        6         Pre        A         SU      1.419178\n",
       "Pre_8  FS09baseline        8         Pre        P         F       1.112329\n",
       "Pre_9  FS10baseline        9         Pre        A         SU      1.389041\n",
       "Pre_10 FS11baseline       10         Pre        P         F       2.191781\n",
       "Pre_11 FS12baseline       11         Pre        A         F       1.567123\n",
       "Pre_12 FS13baseline       12         Pre        P         F       1.032877\n",
       "Pre_13 FS14baseline       13         Pre        P         F             NA\n",
       "Pre_14 FS15baseline       14         Pre        A         F       3.912329\n",
       "Pre_15 FS16baseline       15         Pre        P         F       3.641096\n",
       "Pre_16 FS17baseline       16         Pre        A         F       2.934247\n",
       "Pre_17 FS18baseline       17         Pre        P         F       2.767123\n",
       "Pre_18 FS19baseline       18         Pre        A         SU      2.221918\n",
       "Pre_19 FS20baseline       19         Pre        A         SU      3.638356\n",
       "Pre_22 FS23baseline       22         Pre        P         F       2.495890\n",
       "Pre_23 FS24baseline       23         Pre        A         SU      3.871233\n",
       "Pre_25 FS26baseline       25         Pre        P         F       3.356164\n",
       "Pre_27 FS28baseline       27         Pre        A         SU      3.553425\n",
       "Pre_28 FS29baseline       28         Pre        A         F       3.643836\n",
       "Pre_29 FS30baseline       29         Pre        A         F       3.528767\n",
       "Pre_30 FS31baseline       30         Pre        P         F       1.890411\n",
       "Pre_31 FS32baseline       31         Pre        P         F       1.663014\n",
       "Pre_32 FS33baseline       32         Pre        A         SU      1.663014\n",
       "Pre_34 FS35baseline       34         Pre        A         F       2.972603\n",
       "Pre_35 FS36baseline       35         Pre        P         F       3.030137\n",
       "Pre_36 FS37baseline       36         Pre        A         F       1.290411\n",
       "Pre_37 FS39baseline       37         Pre        P         F       1.715068\n",
       "Pre_39 FS41baseline       39         Pre        A         SU      1.145205\n",
       "Pre_41 FS43baseline       41         Pre        A         SU      2.942466\n",
       "Pre_42 FS44baseline       42         Pre        P         F       1.635616\n",
       "Pre_43 FS45baseline       43         Pre        P         F       3.939726\n",
       "Pre_44 FS46baseline       44         Pre        A         SU      1.769863\n",
       "Pre_45 FS47baseline       45         Pre        P         F       1.769863\n",
       "       IgE    IgE_Binarized\n",
       "Pre_1   41.00 High         \n",
       "Pre_2   14.20 Low          \n",
       "Pre_3   67.40 High         \n",
       "Pre_4    9.59 Low          \n",
       "Pre_5  106.00 High         \n",
       "Pre_6    1.32 Low          \n",
       "Pre_8   19.10 High         \n",
       "Pre_9    0.42 Low          \n",
       "Pre_10   1.74 Low          \n",
       "Pre_11   4.34 Low          \n",
       "Pre_12     NA NA           \n",
       "Pre_13     NA Low          \n",
       "Pre_14 105.00 High         \n",
       "Pre_15   3.70 Low          \n",
       "Pre_16   6.23 Low          \n",
       "Pre_17  18.90 High         \n",
       "Pre_18   4.30 Low          \n",
       "Pre_19   0.31 Low          \n",
       "Pre_22   4.57 Low          \n",
       "Pre_23   1.85 Low          \n",
       "Pre_25  98.70 High         \n",
       "Pre_27   0.83 Low          \n",
       "Pre_28  13.70 Low          \n",
       "Pre_29 180.00 High         \n",
       "Pre_30  96.70 High         \n",
       "Pre_31 113.00 High         \n",
       "Pre_32 113.00 Low          \n",
       "Pre_34 156.00 High         \n",
       "Pre_35  14.10 Low          \n",
       "Pre_36   8.98 Low          \n",
       "Pre_37  19.90 High         \n",
       "Pre_39   5.18 Low          \n",
       "Pre_41   5.61 Low          \n",
       "Pre_42   4.10 Low          \n",
       "Pre_43  71.70 High         \n",
       "Pre_44   2.32 Low          \n",
       "Pre_45   2.32 High         \n",
       "       k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Prevotellaceae.g__Alloprevotella.s__sp._HMT_308\n",
       "Pre_1    5                                                                                                            \n",
       "Pre_2   78                                                                                                            \n",
       "Pre_3    0                                                                                                            \n",
       "Pre_4    0                                                                                                            \n",
       "Pre_5    0                                                                                                            \n",
       "Pre_6    0                                                                                                            \n",
       "Pre_8    0                                                                                                            \n",
       "Pre_9    0                                                                                                            \n",
       "Pre_10   0                                                                                                            \n",
       "Pre_11   0                                                                                                            \n",
       "Pre_12   0                                                                                                            \n",
       "Pre_13   0                                                                                                            \n",
       "Pre_14 327                                                                                                            \n",
       "Pre_15   0                                                                                                            \n",
       "Pre_16  64                                                                                                            \n",
       "Pre_17  42                                                                                                            \n",
       "Pre_18   0                                                                                                            \n",
       "Pre_19   0                                                                                                            \n",
       "Pre_22   0                                                                                                            \n",
       "Pre_23   0                                                                                                            \n",
       "Pre_25   0                                                                                                            \n",
       "Pre_27   8                                                                                                            \n",
       "Pre_28 196                                                                                                            \n",
       "Pre_29  42                                                                                                            \n",
       "Pre_30   0                                                                                                            \n",
       "Pre_31 860                                                                                                            \n",
       "Pre_32   0                                                                                                            \n",
       "Pre_34   0                                                                                                            \n",
       "Pre_35  67                                                                                                            \n",
       "Pre_36   0                                                                                                            \n",
       "Pre_37   0                                                                                                            \n",
       "Pre_39   0                                                                                                            \n",
       "Pre_41   0                                                                                                            \n",
       "Pre_42   0                                                                                                            \n",
       "Pre_43   0                                                                                                            \n",
       "Pre_44   0                                                                                                            \n",
       "Pre_45   0                                                                                                            \n",
       "       k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Porphyromonadaceae.g__Porphyromonas.__\n",
       "Pre_1     0                                                                                                  \n",
       "Pre_2     0                                                                                                  \n",
       "Pre_3     0                                                                                                  \n",
       "Pre_4    13                                                                                                  \n",
       "Pre_5    32                                                                                                  \n",
       "Pre_6   387                                                                                                  \n",
       "Pre_8     0                                                                                                  \n",
       "Pre_9   703                                                                                                  \n",
       "Pre_10    0                                                                                                  \n",
       "Pre_11  215                                                                                                  \n",
       "Pre_12   38                                                                                                  \n",
       "Pre_13   44                                                                                                  \n",
       "Pre_14    0                                                                                                  \n",
       "Pre_15   93                                                                                                  \n",
       "Pre_16   75                                                                                                  \n",
       "Pre_17   88                                                                                                  \n",
       "Pre_18    0                                                                                                  \n",
       "Pre_19  210                                                                                                  \n",
       "Pre_22  911                                                                                                  \n",
       "Pre_23    0                                                                                                  \n",
       "Pre_25    8                                                                                                  \n",
       "Pre_27  151                                                                                                  \n",
       "Pre_28   37                                                                                                  \n",
       "Pre_29  107                                                                                                  \n",
       "Pre_30    0                                                                                                  \n",
       "Pre_31   88                                                                                                  \n",
       "Pre_32 2037                                                                                                  \n",
       "Pre_34  158                                                                                                  \n",
       "Pre_35  145                                                                                                  \n",
       "Pre_36  173                                                                                                  \n",
       "Pre_37  655                                                                                                  \n",
       "Pre_39  268                                                                                                  \n",
       "Pre_41  357                                                                                                  \n",
       "Pre_42  909                                                                                                  \n",
       "Pre_43    0                                                                                                  \n",
       "Pre_44  617                                                                                                  \n",
       "Pre_45  179                                                                                                  \n",
       "       k__Bacteria.p__Fusobacteria.c__Fusobacteriia.o__Fusobacteriales.f__Leptotrichiaceae.g__Pseudoleptotrichia.s__sp._HMT_221\n",
       "Pre_1    23                                                                                                                    \n",
       "Pre_2     0                                                                                                                    \n",
       "Pre_3     0                                                                                                                    \n",
       "Pre_4     9                                                                                                                    \n",
       "Pre_5     0                                                                                                                    \n",
       "Pre_6     0                                                                                                                    \n",
       "Pre_8     0                                                                                                                    \n",
       "Pre_9     0                                                                                                                    \n",
       "Pre_10    9                                                                                                                    \n",
       "Pre_11    0                                                                                                                    \n",
       "Pre_12    0                                                                                                                    \n",
       "Pre_13    0                                                                                                                    \n",
       "Pre_14  575                                                                                                                    \n",
       "Pre_15    0                                                                                                                    \n",
       "Pre_16   13                                                                                                                    \n",
       "Pre_17  115                                                                                                                    \n",
       "Pre_18    0                                                                                                                    \n",
       "Pre_19    0                                                                                                                    \n",
       "Pre_22  117                                                                                                                    \n",
       "Pre_23    0                                                                                                                    \n",
       "Pre_25    0                                                                                                                    \n",
       "Pre_27   59                                                                                                                    \n",
       "Pre_28    0                                                                                                                    \n",
       "Pre_29    4                                                                                                                    \n",
       "Pre_30    0                                                                                                                    \n",
       "Pre_31  360                                                                                                                    \n",
       "Pre_32    0                                                                                                                    \n",
       "Pre_34   14                                                                                                                    \n",
       "Pre_35   86                                                                                                                    \n",
       "Pre_36    0                                                                                                                    \n",
       "Pre_37    0                                                                                                                    \n",
       "Pre_39  116                                                                                                                    \n",
       "Pre_41 1322                                                                                                                    \n",
       "Pre_42    0                                                                                                                    \n",
       "Pre_43    0                                                                                                                    \n",
       "Pre_44 2023                                                                                                                    \n",
       "Pre_45    0                                                                                                                    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "test_bacterium = saliva_raw_model %>%\n",
    "    filter(q_OutcomeSU < 0.1)\n",
    "\n",
    "#ggplot(\n",
    "baseline_saliva_df %>%\n",
    "           select(c(colnames(baseline_saliva_df)[1:8],test_bacterium$taxon))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04b78725-9620-4582-accf-330c3ebd126f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
